{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e338c050",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import seaborn as sns\n",
    "import shutil\n",
    "\n",
    "from sklearn.impute import SimpleImputer\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.metrics import average_precision_score\n",
    "from sklearn.model_selection import cross_val_score\n",
    "import subprocess as sp\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import plot_precision_recall_curve\n",
    "from sklearn import metrics\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "work_dir = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Publications/gene-regulatory-networks-in-epidermal-and-corneal-epithelia/\"\n",
    "os.chdir(f\"{work_dir}/notebooks/Python_functions/\")\n",
    "from Python_scripts import distance_weight\n",
    "from Python_scripts import make_bedfile_from_column\n",
    "from Python_scripts import TSS_window_to_region\n",
    "from Python_scripts import make_autopct\n",
    "from Python_scripts import bedtool_closest\n",
    "\n",
    "plt.style.use(\"classic\")\n",
    "%matplotlib inline\n",
    "#%load_ext nb_black\n",
    "#%reload_ext nb_black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6651bb4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class analysis:\n",
    "    \"\"\"contains all the first steps of my general analysis, including reading a filepath file, config file, and\n",
    "    exporting those and the conda environment to a logdir.\n",
    "    \n",
    "    required input:\n",
    "    datapaths_file, a tsv file containing all the other filepaths needed in the subsequent analysis\n",
    "    config_file, a tsv file containing all type of settings and paramters\n",
    "    output_dir, where to store the output generated by the analysis, it will generate \n",
    "    a log and figure dir in this ouput dir\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, datapaths_file: str, config_file: str, output_dir: str, notebook_file: str):\n",
    "        self.datapaths_file = datapaths_file\n",
    "        self.config_file = config_file\n",
    "        self.output_dir = output_dir\n",
    "        self.files = pd.read_table(\n",
    "            self.datapaths_file, sep=\"\\t\", comment=\"#\", index_col=0, header = 0\n",
    "        )\n",
    "        self.config = pd.read_table(\n",
    "            self.config_file, sep=\"\\t\", comment=\"#\", index_col=0, header = 0\n",
    "        )\n",
    "        self.notebook_file = notebook_file\n",
    "        self.logdir = f\"{self.output_dir}/logs\"\n",
    "        self.figdir = f\"{self.output_dir}/figures\"\n",
    "        self.trackhubdir = f\"{self.output_dir}/trackhub\"\n",
    "        self.tmpdir = f\"{self.output_dir}/tmp\"\n",
    "        Path(self.tmpdir).mkdir(parents=True, exist_ok=True)\n",
    "        Path(self.logdir).mkdir(parents=True, exist_ok=True)\n",
    "        Path(self.figdir).mkdir(parents=True, exist_ok=True)\n",
    "        Path(self.trackhubdir).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "    def save_settings(self, conda_path):\n",
    "        \"\"\"save the config and files to the logs, and make a copy of the conda env that his analysis\n",
    "        is ran in. Takes a full path to the conda env location as input\"\"\"\n",
    "        #save the conda environment\n",
    "        !{conda_path} list > {self.logdir}/conda_env.txt\n",
    "        # save both the files and settings of this run to the output_dir\n",
    "        self.files.T.to_csv(f\"{self.logdir}/filepaths.tsv\", header=False, sep=\"\\t\")\n",
    "        self.config.to_csv(f\"{self.logdir}/config.tsv\", header=False, sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4b756870",
   "metadata": {},
   "outputs": [],
   "source": [
    "MotifEnr = analysis(\n",
    "    datapaths_file=f\"{work_dir}/data/Motif_enrichment_files/files.tsv\",\n",
    "    config_file=f\"{work_dir}/data/Motif_enrichment_files/config.tsv\",\n",
    "    output_dir=f\"{work_dir}/Motif_enrichment/02-05-2022\",\n",
    "    notebook_file=f'{work_dir}/notebooks/2_Motif_enrichment.ipynb',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b59b0e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "TF_motif_file = MotifEnr.files.T.TF2motif[0]\n",
    "TF_motif_file = pd.read_table(TF_motif_file, sep=\"\\t\", comment=\"#\")\n",
    "TF_motif_file[\"Factor\"] = TF_motif_file[\"Factor\"].str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7730b8a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lets load the deseq2 expression data:\n",
    "df_transcriptome = pd.read_table(\n",
    "    MotifEnr.files.T.transcriptome[0],\n",
    "    sep=\";\",\n",
    "    index_col=0,\n",
    "    header = 0,\n",
    ")\n",
    "df_transcriptome.columns = [str(col) + \"_RNAseq\" for col in df_transcriptome.columns]\n",
    "df_transcriptome\n",
    "conditions = [\n",
    "    # (df_transcriptome.BaseMean_RNAseq < 5.0),\n",
    "    (\n",
    "        (df_transcriptome.padj_RNAseq < 0.01)\n",
    "        & (df_transcriptome.log2FoldChange_RNAseq < -0.58)\n",
    "    ),\n",
    "    (\n",
    "        (df_transcriptome.padj_RNAseq < 0.01)\n",
    "        & (df_transcriptome.log2FoldChange_RNAseq > 0.58)\n",
    "    ),\n",
    "]\n",
    "choices = [\"LSC_high\", \"KC_high\"]\n",
    "df_transcriptome[\"prediction_type_RNAseq\"] = np.select(\n",
    "    conditions, choices, default=\"constant\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fa98531c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ammount of TFs expressed and in motif db: 502\n",
      "ammount of TFs differentially expressed and in motif db: 85\n"
     ]
    }
   ],
   "source": [
    "TF_motif_file = pd.read_table(MotifEnr.files.T.TF2motif[0], sep=\"\\t\", comment=\"#\")\n",
    "TF_motif_file[\"Factor\"] = TF_motif_file[\"Factor\"].str.upper()\n",
    "genes_expressed = df_transcriptome[df_transcriptome.baseMean_RNAseq > 10]\n",
    "\n",
    "# merge the motif database with expression data, if a TF is not measured its excluded (halfing the database in size)\n",
    "expressed_TF_motif = TF_motif_file.merge(\n",
    "    genes_expressed, left_on=\"Factor\", right_index= True)\n",
    "\n",
    "expressed_TF_motif.iloc[:, [0, 1, 2, 3]].to_csv(\n",
    "    f\"{MotifEnr.output_dir}/expressedTFs.motif2factors.txt\",\n",
    "    header=True,\n",
    "    sep=\"\\t\",\n",
    "    index=False,\n",
    ")\n",
    "\n",
    "print(\n",
    "    f\"ammount of TFs expressed and in motif db: {len(expressed_TF_motif.Factor.unique())}\"\n",
    ")\n",
    "\n",
    "diff_TF_motif = expressed_TF_motif[\n",
    "    expressed_TF_motif[\"prediction_type_RNAseq\"] != \"constant\"\n",
    "]\n",
    "diff_TF_motif.iloc[:, [0, 1, 2, 3]].to_csv(\n",
    "    f\"{MotifEnr.output_dir}/diffTFs.motif2factors.txt\",\n",
    "    header=True,\n",
    "    sep=\"\\t\",\n",
    "    index=False,\n",
    ")\n",
    "print(\n",
    "    f\"ammount of TFs differentially expressed and in motif db: {len(diff_TF_motif.Factor.unique())}\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bd630939",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Publications/gene-regulatory-networks-in-epidermal-and-corneal-epithelia//Motif_enrichment/02-05-2022/gimme.vertebrate.v5.0.motif2factors.txt'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter pfm database\n",
    "pfm_file = open(MotifEnr.files.T.motif_pwm[0], \"r\")\n",
    "all_motifs_dict = {}\n",
    "pwm_list = []\n",
    "\n",
    "\n",
    "new_motif = \"\"\n",
    "for line in pfm_file:\n",
    "    if line.startswith(\">\"):\n",
    "        if new_motif != \"\":\n",
    "            all_motifs_dict[new_motif] = pwm_list\n",
    "            pwm_list = []\n",
    "        new_motif = line.strip()\n",
    "        new_motif = new_motif.strip(\">\")\n",
    "    else:\n",
    "        pwm_list.append([line.strip()])\n",
    "pfm_file.close()\n",
    "\n",
    "# only expressed TFs\n",
    "expr_motifs_dict = {\n",
    "    k: v for k, v in all_motifs_dict.items() if k in (list(expressed_TF_motif.Motif))\n",
    "}\n",
    "expr_pfm_file = open(\n",
    "    f\"{MotifEnr.output_dir}/expressedTFs.pfm\",\n",
    "    \"w\",\n",
    ")\n",
    "for TF, pwm in expr_motifs_dict.items():\n",
    "    expr_pfm_file.write(f\">{TF}\\n\")\n",
    "    for line in pwm:\n",
    "        expr_pfm_file.write(f'{\"\".join(line)}\\n')\n",
    "expr_pfm_file.close()\n",
    "\n",
    "# only diff TFs\n",
    "diff_motifs_dict = {\n",
    "    k: v for k, v in all_motifs_dict.items() if k in (list(diff_TF_motif.Motif))\n",
    "}\n",
    "diff_pfm_file = open(\n",
    "    f\"{MotifEnr.output_dir}/diffTFs.pfm\",\n",
    "    \"w\",\n",
    ")\n",
    "for TF, pwm in diff_motifs_dict.items():\n",
    "    diff_pfm_file.write(f\">{TF}\\n\")\n",
    "    for line in pwm:\n",
    "        diff_pfm_file.write(f'{\"\".join(line)}\\n')\n",
    "diff_pfm_file.close()\n",
    "\n",
    "#lets also copy a original to the same location as the filtered databases\n",
    "shutil.copyfile(MotifEnr.files.T.motif_pwm[0], f\"{MotifEnr.output_dir}/gimme.vertebrate.v5.0.pfm\")\n",
    "shutil.copyfile(MotifEnr.files.T.TF2motif[0], f\"{MotifEnr.output_dir}/gimme.vertebrate.v5.0.motif2factors.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91e17462",
   "metadata": {},
   "source": [
    "###  after collapsing the TF-motif database, lets perform motif enrichment analysis on  all differential/highly variable regions!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "49079c4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_data = MotifEnr.files.T.CREregion_data[0]\n",
    "region_data = pd.read_table(region_data, sep=\",\", comment=\"#\", header=0, index_col=0)\n",
    "#lets filter for only the differential/highly variable regions:\n",
    "annotation_data = region_data.loc[\n",
    "    :, region_data.columns.str.contains(\"prediction_type\")\n",
    "]\n",
    "# lets make a 200bp window\n",
    "region_data[\"summit\"] = region_data.index\n",
    "region_data[[\"Chrom\", \"ChromStart\", \"ChromEnd\"]] = region_data.summit.str.split(\n",
    "    \"[:-]\", expand=True\n",
    ")\n",
    "region_data[\"ChromStart\"] = region_data[\"ChromStart\"].astype(int) - 100\n",
    "region_data[\"ChromEnd\"] = region_data[\"ChromEnd\"].astype(int) + 100\n",
    "region_data[\"200bpsummit\"] = (\n",
    "    region_data[\"Chrom\"].astype(str)\n",
    "    + \":\"\n",
    "    + region_data[\"ChromStart\"].astype(str)\n",
    "    + \"-\"\n",
    "    + region_data[\"ChromEnd\"].astype(str)\n",
    ")\n",
    "region_data.index = region_data[\"200bpsummit\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fc449ee8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prediction_type_H3K27ac</th>\n",
       "      <th>prediction_type_H3K4me3</th>\n",
       "      <th>prediction_type_ATAC</th>\n",
       "      <th>prediction_type_H3K27me3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>summit_H3K27ac</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1:629078-629278</th>\n",
       "      <td>constant</td>\n",
       "      <td>low_signal</td>\n",
       "      <td>LSC_high</td>\n",
       "      <td>low_signal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:629843-630043</th>\n",
       "      <td>constant</td>\n",
       "      <td>low_signal</td>\n",
       "      <td>constant</td>\n",
       "      <td>low_signal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:630422-630622</th>\n",
       "      <td>constant</td>\n",
       "      <td>low_signal</td>\n",
       "      <td>LSC_high</td>\n",
       "      <td>constant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:633977-634177</th>\n",
       "      <td>LSC_high</td>\n",
       "      <td>low_signal</td>\n",
       "      <td>constant</td>\n",
       "      <td>low_signal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:778592-778792</th>\n",
       "      <td>constant</td>\n",
       "      <td>constant</td>\n",
       "      <td>constant</td>\n",
       "      <td>low_signal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9:137757794-137757994</th>\n",
       "      <td>constant</td>\n",
       "      <td>low_signal</td>\n",
       "      <td>LSC_high</td>\n",
       "      <td>low_signal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9:137814230-137814430</th>\n",
       "      <td>constant</td>\n",
       "      <td>low_signal</td>\n",
       "      <td>LSC_high</td>\n",
       "      <td>constant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9:137987361-137987561</th>\n",
       "      <td>constant</td>\n",
       "      <td>low_signal</td>\n",
       "      <td>constant</td>\n",
       "      <td>constant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9:137990148-137990348</th>\n",
       "      <td>constant</td>\n",
       "      <td>low_signal</td>\n",
       "      <td>KC_high</td>\n",
       "      <td>constant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9:138150188-138150388</th>\n",
       "      <td>constant</td>\n",
       "      <td>KC_high</td>\n",
       "      <td>constant</td>\n",
       "      <td>constant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>124062 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      prediction_type_H3K27ac prediction_type_H3K4me3  \\\n",
       "summit_H3K27ac                                                          \n",
       "1:629078-629278                      constant              low_signal   \n",
       "1:629843-630043                      constant              low_signal   \n",
       "1:630422-630622                      constant              low_signal   \n",
       "1:633977-634177                      LSC_high              low_signal   \n",
       "1:778592-778792                      constant                constant   \n",
       "...                                       ...                     ...   \n",
       "9:137757794-137757994                constant              low_signal   \n",
       "9:137814230-137814430                constant              low_signal   \n",
       "9:137987361-137987561                constant              low_signal   \n",
       "9:137990148-137990348                constant              low_signal   \n",
       "9:138150188-138150388                constant                 KC_high   \n",
       "\n",
       "                      prediction_type_ATAC prediction_type_H3K27me3  \n",
       "summit_H3K27ac                                                       \n",
       "1:629078-629278                   LSC_high               low_signal  \n",
       "1:629843-630043                   constant               low_signal  \n",
       "1:630422-630622                   LSC_high                 constant  \n",
       "1:633977-634177                   constant               low_signal  \n",
       "1:778592-778792                   constant               low_signal  \n",
       "...                                    ...                      ...  \n",
       "9:137757794-137757994             LSC_high               low_signal  \n",
       "9:137814230-137814430             LSC_high                 constant  \n",
       "9:137987361-137987561             constant                 constant  \n",
       "9:137990148-137990348              KC_high                 constant  \n",
       "9:138150188-138150388             constant                 constant  \n",
       "\n",
       "[124062 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e6f86953",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/jsmits/tmp/ipykernel_43191/553585019.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  qnorm_intensities[col1] = df[col1]\n",
      "/scratch/jsmits/tmp/ipykernel_43191/553585019.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  qnorm_intensities[col2] = df[col2]\n"
     ]
    }
   ],
   "source": [
    "# annotation_data\n",
    "diff_annotations = [\"LSC_high\", \"KC_high\"]\n",
    "variable_rows = (\n",
    "    annotation_data.iloc[:, 0].isin(diff_annotations)\n",
    "    | annotation_data.iloc[:, 1].isin(diff_annotations)\n",
    "    | annotation_data.iloc[:, 2].isin(diff_annotations)\n",
    "    | annotation_data.iloc[:, 3].isin(diff_annotations)\n",
    ")\n",
    "var_region_data = region_data.loc[variable_rows.values, :]\n",
    "qnorm_intensities = var_region_data.loc[:, region_data.columns.str.contains(\"qnorm\")]\n",
    "\n",
    "#perform rowmean substraction per histone mod\n",
    "for histmark in [\"H3K27ac\", \"ATAC\", \"H3K4me3\", \"H3K27me3\"]:\n",
    "    col1 = f\"KC_qnorm_{histmark}\"\n",
    "    col2 = f\"LSC_qnorm_{histmark}\"\n",
    "    df = qnorm_intensities[[col1, col2]].astype(float)\n",
    "    df = df.sub(df.mean(axis=1), axis=0)\n",
    "    qnorm_intensities[col1] = df[col1]\n",
    "    qnorm_intensities[col2] = df[col2]\n",
    "    #perform motif enrichment with all diff/variable intensity regions to get the non-redudant motif db\n",
    "    qnorm_intensities[[col1,col2]].to_csv(\n",
    "        f\"{MotifEnr.tmpdir}/{histmark}_int.txt\",\n",
    "        header=True,\n",
    "        sep=\"\\t\",\n",
    "        index=True,\n",
    "    )\n",
    "qnorm_intensities.to_csv(\n",
    "    f\"{MotifEnr.tmpdir}/all_marks_int.txt\",\n",
    "    header=True,\n",
    "    sep=\"\\t\",\n",
    "    index=True,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f60e4802",
   "metadata": {},
   "source": [
    "### Perform motif enrichment with all diff/variable intensity regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b19be46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running maelstrom\n",
      "running maelstrom for gimme.vertebrate.v5.0\n"
     ]
    }
   ],
   "source": [
    "print(f\"running maelstrom\")\n",
    "\n",
    "for TF_db in [\"gimme.vertebrate.v5.0\", \"expressedTFs\", \"diffTFs\"]:\n",
    "    Path(f\"{MotifEnr.output_dir}/maelstrom/{TF_db}/all_marks\").mkdir(\n",
    "        parents=True, exist_ok=True\n",
    "    )\n",
    "    if not os.path.exists(\n",
    "        f\"{MotifEnr.output_dir}/maelstrom/{TF_db}/all_marks/gimme.maelstrom.report.html\"\n",
    "    ):\n",
    "        shutil.rmtree(\n",
    "            \"/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/.cache/gimmemotifs\",\n",
    "            ignore_errors=True,\n",
    "        )\n",
    "        shutil.rmtree(\"/home/jsmits/.config/gimmemotifs/\", ignore_errors=True)\n",
    "        print(f\"running maelstrom for {TF_db}\")\n",
    "        sp.check_call(\n",
    "            f\"nice -15 gimme maelstrom \"\n",
    "            f\"{MotifEnr.tmpdir}/all_marks_int.txt \"\n",
    "            f\"{MotifEnr.files.T.genome[0]} \"\n",
    "            f\"{MotifEnr.output_dir}/maelstrom/{TF_db}/all_marks \"\n",
    "            f\"-p {MotifEnr.output_dir}/{TF_db}.pfm \"\n",
    "            f\"--nthreads 4 \"\n",
    "            f\"2> {MotifEnr.logdir}/maelstrom_{TF_db}_all_marks_log.txt\",\n",
    "            shell=True,\n",
    "        )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
