---
title: "RNAseq_Epithelial_fate_KC_LSC"
author: "Jsmits"
date: "March, 2020"
output:
  html_document: default
  pdf_document: default
---
Conda environment used: 'bulk_rna_seq'
```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=12, fig.height=12) 
```

```{r load_libraries message=FALSE, warning=FALSE}
#Sys.unsetenv("GITHUB_PAT")
#devtools::install_github("JGASmits/deSeq2science")
#install.packages('doMC')
#install.packages('colorRamp2')
#install.packages('amap')

library("DESeq2")
library('devtools')
library('EnhancedVolcano')
library('ComplexHeatmap')
library('cowplot')
library('clusterProfiler')
library('tidyverse')
library('gridExtra')
library('grid')
library('enrichplot')
library('colorRamp2')
library('data.table')
library('readxl')

#BiocManager::install("clusterProfiler ", force = TRUE)#update none

setwd("/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/")
devtools::load_all('deseq2science')
source('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/markdown_notebooks/R_functions/progeny_pathway_enrichment2.R')
```

```{r}
figure_dir <- generate_output_dir('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/FOSL2_knockdown/')
counts <- read.table('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/FOSL2_knockdown/results/counts/GRCh38.p13-counts.tsv', header = T)
Ensemble2Hugo <- read.table('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/genomepy_genomes/GRCh38.p13/gene_id2name.tsv', header= F)
colnames(Ensemble2Hugo) <- c('gene', 'hugo_gene')
counts <- merge(counts,Ensemble2Hugo, by = 'gene' )
counts$gene <- NULL
counts <- counts[!duplicated(counts$hugo_gene),]
rownames(counts) <- counts$hugo_gene
counts$hugo_gene <- NULL
sample_file <- load_s2s_sample_file('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/FOSL2_knockdown/')
sample_file$donor <- as.factor(sample_file$donor)
sample_file$condition <- as.factor(sample_file$condition)
```
Lets vizualize the percentage of knockdown first
```{r}
TPM_df <- read.table('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/FOSL2_knockdown/results/counts/GRCh38.p13-TPM.tsv', header = T)
TPM_df <- merge(TPM_df,Ensemble2Hugo, by = 'gene' )
TPM_df$gene <- NULL
TPM_df <- TPM_df[!duplicated(TPM_df$hugo_gene),]
TPM_df <- TPM_df[!colnames(TPM_df) %in% c('shFOXC1_513','shFOXC1_1094','shFOXC1_1099','shFOXC1_518')]
TPM_df_sub <- TPM_df[TPM_df$hugo_gene %in% c('PAX6','FOSL2'),]

TPM_df_long <- gather(TPM_df_sub, sample, TPM, -hugo_gene)

#add sampel info
TPM_df_long <- merge(TPM_df_long, sample_file, by.x = 'sample', by.y = 'descriptive_name', all.x = TRUE)

pdf(paste0(figure_dir,'/TPM_knockdown_genes.pdf') ,width=12,height=12,paper='special')
for (donor in unique(TPM_df_long$donor)){
  print(donor)
  data_sub <- TPM_df_long[TPM_df_long$donor == donor,]
  p <- ggplot(data = data_sub, aes(x=condition, y = TPM)) + geom_bar(stat = 'identity') +facet_wrap('hugo_gene', scales ='free')  + ggtitle(donor)
  print(p)
}
dev.off()

ggplot(data = TPM_df_long, aes(x=condition, y = TPM, fill = donor)) + geom_bar(stat = 'identity', position = position_dodge()) +facet_wrap('hugo_gene', scales ='free')

#normalize control to 1
for (donor in unique(TPM_df_long$donor)){
  ctr_col <- paste0('shCTR_',str_sub(donor,2))
  Pax6_col <- paste0('shPAX6_',str_sub(donor,2))
  fosl2_col <- paste0('shFOSL2_',str_sub(donor,2))
  if (fosl2_col %in% colnames(TPM_df_sub)){
    print(paste0(donor, ' fosl2'))
    TPM_df_sub[fosl2_col] = TPM_df_sub[fosl2_col] / TPM_df_sub[ctr_col]*100
    TPM_df_sub[ctr_col]= TPM_df_sub[ctr_col] / TPM_df_sub[ctr_col]*100
  }
  if (Pax6_col %in% colnames(TPM_df_sub)){
    print(paste0(donor, ' PAX6'))
    TPM_df_sub[Pax6_col] = TPM_df_sub[Pax6_col] / TPM_df_sub[ctr_col] *100
  TPM_df_sub[ctr_col]= TPM_df_sub[ctr_col] / TPM_df_sub[ctr_col] *100
  }
}
sample_file$donor_sample <- paste(sample_file$condition ,sample_file$donor)

TPM_df_long <- gather(TPM_df_sub, sample, TPM, -hugo_gene)

#add sampel info
TPM_df_long <- merge(TPM_df_long, sample_file, by.x = 'sample', by.y = 'descriptive_name', all.x = TRUE)

pdf(paste0(figure_dir,'/TPM_knockdown_genes_together.pdf') ,width=8,height=4,paper='special')
ggplot(data = TPM_df_long, aes(x=donor, y = TPM, fill = sample)) + geom_bar(stat = 'identity', position = position_dodge()) +facet_wrap('hugo_gene')
dev.off()
```
```{r}
TPM_df_sub <- TPM_df[TPM_df$hugo_gene %in% c('ATF2','ATF3','ATF4','BATF','BATF3','FOS','FOSL1','FOSL2','FOSB','MAF', 'MAFA','MAFB','MAFG','JUN','JUNB','JUND'),]

TPM_df_long <- gather(TPM_df_sub, sample, TPM, -hugo_gene)

#add sampel info
TPM_df_long <- merge(TPM_df_long, sample_file, by.x = 'sample', by.y = 'descriptive_name', all.x = TRUE)

#normalize control to 1
for (donor in unique(TPM_df_long$donor)){
  ctr_col <- paste0('shCTR_',str_sub(donor,2))
  Pax6_col <- paste0('shPAX6_',str_sub(donor,2))
  fosl2_col <- paste0('shFOSL2_',str_sub(donor,2))
  if (fosl2_col %in% colnames(TPM_df_sub)){
    print(paste0(donor, ' fosl2'))
    TPM_df_sub[fosl2_col] = TPM_df_sub[fosl2_col] / TPM_df_sub[ctr_col]*100
    TPM_df_sub[ctr_col]= TPM_df_sub[ctr_col] / TPM_df_sub[ctr_col]*100
  }
  if (Pax6_col %in% colnames(TPM_df_sub)){
    print(paste0(donor, ' PAX6'))
    TPM_df_sub[Pax6_col] = TPM_df_sub[Pax6_col] / TPM_df_sub[ctr_col] *100
  TPM_df_sub[ctr_col]= TPM_df_sub[ctr_col] / TPM_df_sub[ctr_col] *100
  }
}
sample_file$donor_sample <- paste(sample_file$condition ,sample_file$donor)

TPM_df_long <- gather(TPM_df_sub, sample, TPM, -hugo_gene)

#add sampel info
TPM_df_long <- merge(TPM_df_long, sample_file, by.x = 'sample', by.y = 'descriptive_name', all.x = TRUE)

#pdf(paste0(figure_dir,'/TPM_knockdown_genes_together.pdf') ,width=8,height=4,paper='special')
ggplot(data = TPM_df_long, aes(x=donor, y = TPM, fill = sample)) + geom_bar(stat = 'identity', position = position_dodge()) +facet_wrap('hugo_gene', scales = 'free')
#dev.off()
```


Lets swap the counts
```{r}
#sample_file <- sample_file[!sample_file$donor == 'd518',]
sample_file_PAX6 <- sample_file[sample_file$donor %in% c('d868','d874','d888'),]
sample_file_FOSL2 <- sample_file[!sample_file$donor %in% c('d868','d874','d888'),]
sample_file_FOSL2 <- sample_file_FOSL2[!sample_file_FOSL2$condition %in% c('shFOXC1'),]
sample_file_FOSL2$condition
counts_PAX6 <- counts[,colnames(counts) %in% sample_file_PAX6$descriptive_name]
count_FOSL2 <- counts[,colnames(counts) %in% sample_file_FOSL2$descriptive_name]

dds_PAX6 <- DESeq2::DESeqDataSetFromMatrix(countData = counts_PAX6,
                          #generate deseq object, the samples and designs are specified in deseq_objects_df
                          colData =sample_file_PAX6,
                          design = as.formula( ~ donor + condition))
dds_FOSL2 <- DESeq2::DESeqDataSetFromMatrix(countData = count_FOSL2,
                          #generate deseq object, the samples and designs are specified in deseq_objects_df
                          colData =sample_file_FOSL2,
                          design = as.formula( ~ donor + condition))
```

```{r}
keep <- rowSums(DESeq2::counts(dds_PAX6)) >= 10
dds_PAX6 <- dds_PAX6[keep,]
#set all as the baseline
dds_PAX6[['condition']] <- relevel(dds_PAX6[['condition']],'shCTR')
dds_PAX6 <- DESeq(dds_PAX6)
#make a list of all genes measured with more than 10 counts to use as a background.
gene_background_list_PAX6 <- rownames(DESeq2::counts(dds_PAX6)[rowSums(DESeq2::counts(dds_PAX6)) > 10,])
rlog_PAX6 <- DESeq2::vst(dds_PAX6) #normalize deseq2 intensities

dds_PAX6$sample
```
```{r}
keep <- rowSums(DESeq2::counts(dds_FOSL2)) >= 10
dds_FOSL2 <- dds_FOSL2[keep,]
#set all as the baseline
dds_FOSL2[['condition']] <- relevel(dds_FOSL2[['condition']],'shCTR')
dds_FOSL2 <- DESeq(dds_FOSL2)
#make a list of all genes measured with more than 10 counts to use as a background.
gene_background_list_FOSL2 <- rownames(DESeq2::counts(dds_FOSL2)[rowSums(DESeq2::counts(dds_FOSL2)) > 10,])
rlog_FOSL2 <- DESeq2::vst(dds_FOSL2) #normalize deseq2 intensities
```


```{r}
plot_dist_dds(rlog_PAX6,paste0(figure_dir,"/distance_PAX6.pdf"),"donor","condition")
ammount_of_sig_PCs <- plotPCA_dds(rlog_PAX6,
                     intgroup = c('donor','condition'),
                     output_dir =figure_dir,
                     filename= 'PAX6_PCA_',
                     run_go_enrichment = T,
                     Genenames_keyType = 'SYMBOL',
                     n_PCA_top_genes = 15,
                     ntop = 1000,
                     PC_cutoff = 5,
                     #gene_background = gene_background_list_PAX6
)
plot_dist_dds(rlog_FOSL2,paste0(figure_dir,"/distance_FOSL2.pdf"),"donor","condition")
ammount_of_sig_PCs <- plotPCA_dds(rlog_FOSL2,
                     intgroup = c('donor','condition'),
                     output_dir =figure_dir,
                     filename= 'FOSL2_PCA_',
                     run_go_enrichment = T,
                     Genenames_keyType = 'SYMBOL',
                     n_PCA_top_genes = 15,
                     ntop = 1000,
                     PC_cutoff = 5,
                     #gene_background = gene_background_list_FOSL2
)
```

```{r}
comparison_FOSL2 = list(
  #wild type differentiation DEGS
  c('condition','shFOSL2','shCTR')
  #c('condition','shFOXC1','shCTR')
)

comparison_PAX6 = list(
  #wild type differentiation DEGS
  c('condition','shPAX6','shCTR')
  #c('condition','shFOXC1','shCTR')
)
```
#lets plot the percentage reduction per donor
```{r}
# highlight_genes <- c('FOSL2', 'PAX6', 'FOXC1')
# 
# tcounts <-
#   t(log2((
#     counts(dds[highlight_genes, ], normalized = TRUE, replaced = FALSE) + .5
#   ))) %>%
#   merge(colData(dds), ., by = "row.names") %>%
#   tidyr::gather(gene, expression, (ncol(.) - length(highlight_genes) + 1):ncol(.))
# donor = 'd1094'
# 
# for (donor in unique(tcounts$donor)) {
#   print(donor)
#   counts_donor <- tcounts[tcounts$donor == donor,]
#   for (gene in unique(counts_donor$gene)) {
#     counts_donor_gene <- counts_donor[counts_donor$gene == gene,]
#     counts_donor_gene <-
#       counts_donor_gene[(
#         counts_donor_gene$condition == paste0('sh', gene) |
#           counts_donor_gene$condition == 'shCTR'
#       ),]
#     if (nrow(counts_donor_gene) > 1) {
#       CTR_counts <-
#         2 ^ counts_donor_gene[counts_donor_gene$condition == 'shCTR',]$expression
#       sh_counts <-
#         
#         
#         
#         2 ^ counts_donor_gene[!counts_donor_gene$condition == 'shCTR',]$expression
#       FC_percentage <- sh_counts / CTR_counts * 100
#       print(paste0(gene, ' ', FC_percentage))
#     }
#   }
# }
```


```{r}
DEG_pval_cutoff <- 0.05
DEG_log2FC_cutoff <- 0.58
all_DEGs <- list()
#run DEG analysis for each comparison
comparisons <- comparison_FOSL2
dds <- dds_FOSL2
gene_background_list <- gene_background_list_FOSL2
rlog <- rlog_FOSL2
vulcano_genes <- c('FOSL2','JAG1','ALDH1A1','CCT8','VEGFA','TGFBI','VIM','SNAI2','SMAD3','SERPIN13','CLDN4','TGM1')

metacolumn <- comparisons[[1]][1]
condition1 <- comparisons[[1]][3]
condition2 <- comparisons[[1]][2]
coef <- paste0(metacolumn, '_', condition2, '_vs_', condition1)
print(coef)
dir_name <- paste0(condition1, '_negFC vs ', condition2, '_posFC')
#make an output dir for all the files and analysis for each comparison
comparison_combi_dir <- paste(figure_dir , dir_name, sep = '/')
dir.create(file.path(comparison_combi_dir), showWarnings = FALSE)
if (metacolumn == 'condition') {
  if (!coef %in% resultsNames(dds)) {
    dds[[metacolumn]] <-
      relevel(dds[[metacolumn]], ref = condition1)#relevel object to condition 1
    dds <-
      nbinomWaldTest(dds)#rerun the nbegative binomal wald test to enable apeglm shrinkage for specific comparison
  }
  res <- lfcShrink(dds, coef = coef, type = "apeglm")
  write.table(
    as.data.frame(res),
    file = paste(comparison_combi_dir, "/all_genes.csv", sep = ""),
    sep = ','
  )
  #filter DEGS
  diff_genes <- subset(res, padj < as.numeric(DEG_pval_cutoff))
  diff_genes <-
    subset(
      diff_genes,
      log2FoldChange > as.numeric(DEG_log2FC_cutoff) |
        log2FoldChange < -1 * as.numeric(DEG_log2FC_cutoff)
    )
  diff_genes <- as.data.frame(diff_genes)
  write.table(
    as.data.frame(diff_genes),
    file = paste(comparison_combi_dir, "/DEGs.csv", sep = ""),
    sep = ';',
    row.names = T
  )
  all_DEGs <-
    append(all_DEGs, row.names(diff_genes)) #add all DEG genes
  
  DEGS_up <- diff_genes[diff_genes$log2FoldChange > 0, ]
  DEGS_dwn <- diff_genes[diff_genes$log2FoldChange < 0, ]
  
  #make a volcano plot, blegh
  vulcano_plot <- EnhancedVolcano(
    res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue',
    title = dir_name,
    pCutoff = as.numeric(DEG_pval_cutoff),
    FCcutoff = as.numeric(DEG_log2FC_cutoff),
    xlab = bquote( ~ Log[2] ~ 'fold change'),
    pointSize = 4.0,
    labSize = 6.0,
    colAlpha = 1,
    legendPosition = 'right',
    legendLabSize = 12,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 0.75
  )
    vulcano_plot_selected <- EnhancedVolcano(
    res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue',
    selectLab = vulcano_genes,
    title = dir_name,
    pCutoff = as.numeric(DEG_pval_cutoff),
    FCcutoff = as.numeric(DEG_log2FC_cutoff),
    xlab = bquote( ~ Log[2] ~ 'fold change'),
    pointSize = 4.0,
    labSize = 6.0,
    colAlpha = 1,
    legendPosition = 'right',
    legendLabSize = 12,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 0.75
  )
  
  
  pdf(
    paste0(comparison_combi_dir, '/Differential_genes_vulcano_labels.pdf') ,
    width = 12,
    height = 12,
    paper = 'special'
  )
  print(vulcano_plot)
  print(vulcano_plot_selected)
  dev.off()
  #run GO term enrichment for the DEGS
  if (!file.exists(paste0(comparison_combi_dir, '/GO_all_DEGS.csv'))) {
    GOenr_simplified(
      gene_list = rownames(diff_genes),
      output_file = paste0(comparison_combi_dir, '/GO_all_DEGS'),
      res,
      dds,
      fill_annotation = 'donor',
      #gene_background_list = gene_background_list,
      x_axis_annotation = 'condition',
      simplify_value = 0.6
    )
  }
  if (!file.exists(paste0(
    comparison_combi_dir,
    '/GO_DEGS_high_',
    condition2,
    '.csv'
  ))) {
    GOenr_simplified(
      gene_list = rownames(DEGS_up),
      output_file = paste0(comparison_combi_dir, '/GO_DEGS_high_', condition2),
      res,
      gene_background_list = gene_background_list,
      dds,
      fill_annotation = 'donor',
      x_axis_annotation = 'condition',
      simplify_value = 0.6
    )
  }
  if (!file.exists(paste0(
    comparison_combi_dir,
    '/GO_DEGS_high_',
    condition1,
    '.csv'
  ))) {
    GOenr_simplified(
      gene_list = rownames(DEGS_dwn),
      output_file = paste0(comparison_combi_dir, '/GO_DEGS_high', condition1),
      res,
      gene_background_list = gene_background_list,
      dds,
      fill_annotation = 'donor',
      x_axis_annotation = 'condition',
      simplify_value = 0.6
    )
  }
  #generate a heatmap of the differential genes
  output_filename <-
    paste0(comparison_combi_dir, '/DEG_heatmap.pdf')
  if (!file.exists(output_filename)) {
    mat <- assay(rlog)[row.names(diff_genes),]
    mat <- t(scale(t(mat)))
    if (nrow(mat) < 100) {
      rowname_var <- TRUE
    } else{
      rowname_var <- FALSE
    }
    
    ha = ComplexHeatmap::columnAnnotation(annotation1 = rlog[['condition']],
                                          annotation2 = rlog[['donor']])
    RNA_heatmap <- ComplexHeatmap::Heatmap(
      mat,
      cluster_columns = T,
      cluster_rows = T,
      show_row_names = rowname_var,
      top_annotation = ha,
      row_km = 2
    )
    pdf(
      paste(comparison_combi_dir, '/DEG_heatmap.pdf', sep = "") ,
      width = 8,
      height = 8,
      paper = 'special'
    )
    print(RNA_heatmap)
    dev.off()
  }
  GSEA_mysigdb(res,
               comparison_combi_dir,
               ncores = 1)
  
  GSEA_list_dir <-
    '/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/FOSL2_knockdown/costum_gene_sets/'
  all_gene_sets <- list.files(path = GSEA_list_dir)
  
  GSE_targets <- NULL
  for (geneset in all_gene_sets) {
    #check if filesize is not 0
    if (file.info(paste(GSEA_list_dir, geneset, sep = '/'))$size > 0) {
      print(geneset)
      if (is.null(GSE_targets)) {
        print(geneset)
        geneset_genes <-
          read.csv2(paste(GSEA_list_dir, geneset, sep = '/'), header = F)
        GSE_targets <- data.frame("term" = geneset,
                                  "gene" = unique(geneset_genes))
      }
      else{
        geneset_genes <-
          read.csv2(paste(GSEA_list_dir, geneset, sep = '/'), header = F)
        GSE_targets2 <- data.frame("term" = geneset,
                                   "gene" = unique(geneset_genes))
        GSE_targets <- rbind(GSE_targets, GSE_targets2)
      }
      
    }
  }
  #Run GSEA on the comparison
  #lets load the GSEA objects from the GSEA containing DIR
  print('running GSEA')
  #Perform Gene Set Enrichment analysis for a specific Gene Set
  GSEA_dir <- paste0(comparison_combi_dir, '/GSEA')
  GSSEA_plots <-  paste0(GSEA_dir, '/costum_db')
  dir.create(file.path(GSEA_dir), showWarnings = FALSE)
  dir.create(file.path(GSSEA_plots), showWarnings = FALSE)
  
  #lets make a ranked genelist
  res_GSEA <- as.data.frame(res)
  res_GSEA <- drop_na(res_GSEA)
  FC_list <- res_GSEA$log2FoldChange
  names(FC_list) <- as.character(rownames(res_GSEA))
  geneList <- sort(FC_list, decreasing = TRUE)
  
  GSEA_results <-
    GSEA(
      geneList,
      TERM2GENE = GSE_targets,
      pvalueCutoff = 1,
      eps = 0,
      maxGSSize = 10000
    )
  #export the GSEA result
  write.table(
    as.data.frame(GSEA_results),
    row.names = T,
    col.names = T,
    file = paste0(GSEA_dir, "/GSEA_results.csv"),
    sep = ','
  )
  GSEA_df <- as.data.frame(GSEA_results)
  GSEA_df <- GSEA_df[c('Description', 'enrichmentScore', 'qvalues')]
  #if the costum GSEA list is bellow 5 length, lets plot all of them in a single graph:
  if (length(as.data.frame(GSEA_results@result)[1]$ID) < 6) {
    print("costum GSEA db < 5 gene sets, plotting all in one graph")
    GSEA_plot <-
      gseaplot2(GSEA_results ,
                geneSetID = 1:(length(
                  as.data.frame(GSEA_results@result)[1]$ID
                )),
                pvalue_table = T)
    plotname <- paste0('all_costum_genesets_enrichment.pdf')
    pdf(
      paste(GSSEA_plots, plotname, sep = "/"),
      width = 8,
      height = 6,
      paper = 'special'
    )
    print(GSEA_plot)
    dev.off()
  }
  
  #also plot the top 10 GSEA seperate
  
  gseares = 1
  for (sigGSEA in as.data.frame(GSEA_results@result)[1]$ID) {
    GSEA_plot <-
      gseaplot2(GSEA_results ,
                geneSetID = gseares,
                pvalue_table = T)
    plotname <- paste0(sigGSEA, '_GSEA_enrichment.pdf')
    pdf(
      paste(GSSEA_plots, plotname, sep = "/"),
      width = 8,
      height = 6,
      paper = 'special'
    )
    print(GSEA_plot)
    gseares = gseares + 1
    dev.off()
  }
}

output_filename <- paste(comparison_combi_dir,'/PROGENy_barplot_analysis.pdf',sep="")
output_filename2 <- paste(comparison_combi_dir,'/progeny_data.csv',sep="")

dds$condition
if(!file.exists(output_filename)){
  print('running progeny pathway enrichiment')
  Progeny_enrichment(rlog, 
                     annotation_column_dds = 'condition', 
                     control_annotation = condition1, 
                     comp_annotation = condition2, 
                     output_pdf = output_filename, 
                     output_matrix = output_filename2)
    # vizualize how the gene expression log2FC is behaving compared to the progeny coef numbers:
  progeny_coef <- getModel(organism = 'Human', top = 200)
  progeny_coef <-
    progeny_coef[apply(progeny_coef[,-1], 1, function(x)
      ! all(x == 0)),]
  #rename JAK-stat
  progeny_coef$JAKSTAT <- progeny_coef$`JAK-STAT`
  progeny_coef$`JAK-STAT` <- NULL
  #lets link the progeny scores to the result object:
  progeny_res <-
    merge(progeny_coef,
          as.data.frame(res),
          by = 0,
          all = TRUE)
  rownames(progeny_res) <- progeny_res$Row.names
  output_filename <-
    paste(comparison_combi_dir, '/progeny_genes.pdf', sep = "")
  
  
  pdf(
    output_filename ,
    width = 8,
    height = 8,
    paper = 'special'
  )
  for (pathway in colnames(progeny_coef)) {
    print(pathway)
    point_plot <- NULL
    subset_Df <- drop_na(progeny_res[c(pathway, 'log2FoldChange')])
    subset_Df <- subset_Df[subset_Df[pathway] != 0,]
    df <-
      gather(subset_Df, category, value, c(-pathway, -'log2FoldChange'))
    df <- mutate_all(df, function(x)
      as.numeric(as.character(x)))
    df$Gene <- rownames(df)
    point_plot <-
      ggplot(df, aes_string(x = colnames(df)[1], y = colnames(df)[2])) +
      geom_point(size = 3)  +
      scale_color_discrete(name = NULL) +
      theme(legend.position = "top")
    point_plot <- point_plot +   geom_label_repel(
      aes(label = Gene),
      box.padding   = 0.35,
      point.padding = 0.5,
      segment.color = 'grey50'
    )
    print(point_plot)
  }
  dev.off()
}
```
FOSL2 Heatmap

```{r}
highlight_genes <- c('TGFBI','FOSL2','JAG1','JUN','RBPJ','CLDN1','TGM1','COL1A1','DLL1','COL8A1','FZD5','MAF','CBX5','ADAM19','DICER1','OCLN','CLDN4','CLDN1','SULT2B1','CGN','TP63','OCLN')

mat <- assay(rlog)[row.names(diff_genes),]
mat <- t(scale(t(mat)))
if (nrow(mat) < 100) {
  rowname_var <- TRUE
} else{
  rowname_var <- FALSE
}
gene_rows <-
  as.numeric(sapply(highlight_genes, function(x)
    which(rownames(mat) == x)))

row_ha = ComplexHeatmap::rowAnnotation(selected_genes = ComplexHeatmap::anno_mark(at = gene_rows, labels = highlight_genes, labels_gp(fontsize = 40)))

ha = ComplexHeatmap::columnAnnotation(annotation1 = rlog[['condition']],
                                      annotation2 = rlog[['donor']])


RNA_heatmap <- ComplexHeatmap::Heatmap(
  mat,
  cluster_columns = T,
  cluster_rows = T,
  show_row_names = rowname_var,
  top_annotation = ha,
  right_annotation = row_ha,
  row_km = 2
)
pdf(
  paste(comparison_combi_dir, '/DEG_heatmap_annotated.pdf', sep = "") ,
  width = 5,
  height = 6,
  paper = 'special'
)

print(RNA_heatmap)
dev.off()
```
GO-enrichment
```{r}
library(clusterProfiler)
GO_obj_up <-
  clusterProfiler::enrichGO(
    gene = rownames(DEGS_up),
    universe = gene_background_list,
    keyType = 'SYMBOL',
    ont = "BP",
    OrgDb = org.Hs.eg.db::org.Hs.eg.db,
    qvalueCutoff = 0.01
  )
GO_obj_dwn <-
  clusterProfiler::enrichGO(
    gene = rownames(DEGS_dwn),
    universe = gene_background_list,
    keyType = 'SYMBOL',
    ont = "BP",
    OrgDb = org.Hs.eg.db::org.Hs.eg.db
  )
View(as.data.frame(GO_obj_up))

go_obj_up_s <- clusterProfiler::simplify(GO_obj_up, cutoff = 0.7)
View(as.data.frame(go_obj_up_s))

pdf(
paste(comparison_combi_dir, '/GO_upDEGS.pdf', sep = "") ,
  width = 6,
  height = 5,
  paper = 'special'
)
clusterProfiler::dotplot(go_obj_up_s, showCategory = 20)
dev.off()
# 
# 
# 
#   GO_up <- GOenr_simplified(
#       gene_list = rownames(DEGS_up),
#       output_file = paste0(comparison_combi_dir, '/GO_DEGS_high', condition1),
#       res,
#       gene_background_list = gene_background_list,
#       dds,
#       fill_annotation = 'donor',
#       x_axis_annotation = 'condition',
#       simplify_value = 0.6
```


```{r}
DEG_pval_cutoff <- 0.05
DEG_log2FC_cutoff <- 0.58
all_DEGs <- list()
#run DEG analysis for each comparison
comparisons <- comparison_PAX6
dds <- dds_PAX6
gene_background_list <- gene_background_list_PAX6
rlog <- rlog_PAX6

metacolumn <- comparisons[[1]][1]
condition1 <- comparisons[[1]][3]
condition2 <- comparisons[[1]][2]
coef <- paste0(metacolumn, '_', condition2, '_vs_', condition1)
print(coef)
dir_name <- paste0(condition1, '_negFC vs ', condition2, '_posFC')
#make an output dir for all the files and analysis for each comparison
comparison_combi_dir <- paste(figure_dir , dir_name, sep = '/')
dir.create(file.path(comparison_combi_dir), showWarnings = FALSE)
if (metacolumn == 'condition') {
  if (!coef %in% resultsNames(dds)) {
    dds[[metacolumn]] <-
      relevel(dds[[metacolumn]], ref = condition1)#relevel object to condition 1
    dds <-
      nbinomWaldTest(dds)#rerun the nbegative binomal wald test to enable apeglm shrinkage for specific comparison
  }
  res <- lfcShrink(dds, coef = coef, type = "apeglm")
  write.table(
    as.data.frame(res),
    file = paste(comparison_combi_dir, "/all_genes.csv", sep = ""),
    sep = ','
  )
  #filter DEGS
  diff_genes <- subset(res, padj < as.numeric(DEG_pval_cutoff))
  diff_genes <-
    subset(
      diff_genes,
      log2FoldChange > as.numeric(DEG_log2FC_cutoff) |
        log2FoldChange < -1 * as.numeric(DEG_log2FC_cutoff)
    )
  diff_genes <- as.data.frame(diff_genes)
  write.table(
    as.data.frame(diff_genes),
    file = paste(comparison_combi_dir, "/DEGs.csv", sep = ""),
    sep = ';',
    row.names = T
  )
  all_DEGs <-
    append(all_DEGs, row.names(diff_genes)) #add all DEG genes
  
  DEGS_up <- diff_genes[diff_genes$log2FoldChange > 0, ]
  DEGS_dwn <- diff_genes[diff_genes$log2FoldChange < 0, ]
  
  #make a volcano plot, blegh
  vulcano_plot <- EnhancedVolcano(
    res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue',
    title = dir_name,
    pCutoff = as.numeric(DEG_pval_cutoff),
    FCcutoff = as.numeric(DEG_log2FC_cutoff),
    xlab = bquote( ~ Log[2] ~ 'fold change'),
    pointSize = 4.0,
    labSize = 6.0,
    colAlpha = 1,
    legendPosition = 'right',
    legendLabSize = 12,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 0.75
  )
  
  pdf(
    paste0(comparison_combi_dir, '/Differential_genes_vulcano.pdf') ,
    width = 12,
    height = 12,
    paper = 'special'
  )
  print(vulcano_plot)
  dev.off()
  #run GO term enrichment for the DEGS
  if (!file.exists(paste0(comparison_combi_dir, '/GO_all_DEGS.csv'))) {
    GOenr_simplified(
      gene_list = rownames(diff_genes),
      output_file = paste0(comparison_combi_dir, '/GO_all_DEGS'),
      res,
      dds,
      fill_annotation = 'donor',
      gene_background_list = gene_background_list,
      x_axis_annotation = 'condition',
      simplify_value = 0.6
    )
  }
  if (!file.exists(paste0(
    comparison_combi_dir,
    '/GO_DEGS_high_',
    condition2,
    '.csv'
  ))) {
    GOenr_simplified(
      gene_list = rownames(DEGS_up),
      output_file = paste0(comparison_combi_dir, '/GO_DEGS_high_', condition2),
      res,
      gene_background_list = gene_background_list,
      dds,
      fill_annotation = 'donor',
      x_axis_annotation = 'condition',
      simplify_value = 0.6
    )
  }
  if (!file.exists(paste0(
    comparison_combi_dir,
    '/GO_DEGS_high_',
    condition1,
    '.csv'
  ))) {
    GOenr_simplified(
      gene_list = rownames(DEGS_dwn),
      output_file = paste0(comparison_combi_dir, '/GO_DEGS_high', condition1),
      res,
      gene_background_list = gene_background_list,
      dds,
      fill_annotation = 'donor',
      x_axis_annotation = 'condition',
      simplify_value = 0.6
    )
  }
  #generate a heatmap of the differential genes
  output_filename <-
    paste0(comparison_combi_dir, '/DEG_heatmap.pdf')
  if (!file.exists(output_filename)) {
    mat <- assay(rlog)[row.names(diff_genes),]
    mat <- t(scale(t(mat)))
    if (nrow(mat) < 100) {
      rowname_var <- TRUE
    } else{
      rowname_var <- FALSE
    }
    
    ha = ComplexHeatmap::columnAnnotation(annotation1 = rlog[['condition']],
                                          annotation2 = rlog[['donor']])
    RNA_heatmap <- ComplexHeatmap::Heatmap(
      mat,
      cluster_columns = T,
      cluster_rows = T,
      show_row_names = rowname_var,
      top_annotation = ha,
      row_km = 4
    )
    pdf(
      paste(comparison_combi_dir, '/DEG_heatmap.pdf', sep = "") ,
      width = 8,
      height = 8,
      paper = 'special'
    )
    print(RNA_heatmap)
    dev.off()
  }
  GSEA_mysigdb(res,
               comparison_combi_dir,
               ncores = 1)
  
  GSEA_list_dir <-
    '/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/FOSL2_knockdown/costum_gene_sets/'
  all_gene_sets <- list.files(path = GSEA_list_dir)
  
  GSE_targets <- NULL
  for (geneset in all_gene_sets) {
    #check if filesize is not 0
    if (file.info(paste(GSEA_list_dir, geneset, sep = '/'))$size > 0) {
      print(geneset)
      if (is.null(GSE_targets)) {
        print(geneset)
        geneset_genes <-
          read.csv2(paste(GSEA_list_dir, geneset, sep = '/'), header = F)
        GSE_targets <- data.frame("term" = geneset,
                                  "gene" = unique(geneset_genes))
      }
      else{
        geneset_genes <-
          read.csv2(paste(GSEA_list_dir, geneset, sep = '/'), header = F)
        GSE_targets2 <- data.frame("term" = geneset,
                                   "gene" = unique(geneset_genes))
        GSE_targets <- rbind(GSE_targets, GSE_targets2)
      }
      
    }
  }
  #Run GSEA on the comparison
  #lets load the GSEA objects from the GSEA containing DIR
  print('running GSEA')
  #Perform Gene Set Enrichment analysis for a specific Gene Set
  GSEA_dir <- paste0(comparison_combi_dir, '/GSEA')
  GSSEA_plots <-  paste0(GSEA_dir, '/costum_db')
  dir.create(file.path(GSEA_dir), showWarnings = FALSE)
  dir.create(file.path(GSSEA_plots), showWarnings = FALSE)
  
  #lets make a ranked genelist
  res_GSEA <- as.data.frame(res)
  res_GSEA <- drop_na(res_GSEA)
  FC_list <- res_GSEA$log2FoldChange
  names(FC_list) <- as.character(rownames(res_GSEA))
  geneList <- sort(FC_list, decreasing = TRUE)
  
  GSEA_results <-
    GSEA(
      geneList,
      TERM2GENE = GSE_targets,
      pvalueCutoff = 1,
      eps = 0,
      maxGSSize = 10000
    )
  #export the GSEA result
  write.table(
    as.data.frame(GSEA_results),
    row.names = T,
    col.names = T,
    file = paste0(GSEA_dir, "/GSEA_results.csv"),
    sep = ','
  )
  GSEA_df <- as.data.frame(GSEA_results)
  GSEA_df <- GSEA_df[c('Description', 'enrichmentScore', 'qvalues')]
  #if the costum GSEA list is bellow 5 length, lets plot all of them in a single graph:
  if (length(as.data.frame(GSEA_results@result)[1]$ID) < 6) {
    print("costum GSEA db < 5 gene sets, plotting all in one graph")
    GSEA_plot <-
      gseaplot2(GSEA_results ,
                geneSetID = 1:(length(
                  as.data.frame(GSEA_results@result)[1]$ID
                )),
                pvalue_table = T)
    plotname <- paste0('all_costum_genesets_enrichment.pdf')
    pdf(
      paste(GSSEA_plots, plotname, sep = "/"),
      width = 8,
      height = 6,
      paper = 'special'
    )
    print(GSEA_plot)
    dev.off()
  }
  
  #also plot the top 10 GSEA seperate
  
  gseares = 1
  for (sigGSEA in as.data.frame(GSEA_results@result)[1]$ID) {
    GSEA_plot <-
      gseaplot2(GSEA_results ,
                geneSetID = gseares,
                pvalue_table = T)
    plotname <- paste0(sigGSEA, '_GSEA_enrichment.pdf')
    pdf(
      paste(GSSEA_plots, plotname, sep = "/"),
      width = 8,
      height = 6,
      paper = 'special'
    )
    print(GSEA_plot)
    gseares = gseares + 1
    dev.off()
  }
}

output_filename <- paste(comparison_combi_dir,'/PROGENy_barplot_analysis.pdf',sep="")
output_filename2 <- paste(comparison_combi_dir,'/progeny_data.csv',sep="")

if (!file.exists(output_filename)) {
  print('running progeny pathway enrichiment')
  Progeny_enrichment(
    rlog,
    annotation_column_dds = 'condition',
    control_annotation = condition1,
    comp_annotation = condition2,
    output_pdf = output_filename,
    output_matrix = output_filename2
  )
  
  
  # vizualize how the gene expression log2FC is behaving compared to the progeny coef numbers:
  progeny_coef <- getModel(organism = 'Human', top = 200)
  progeny_coef <-
    progeny_coef[apply(progeny_coef[,-1], 1, function(x)
      ! all(x == 0)),]
  #rename JAK-stat
  progeny_coef$JAKSTAT <- progeny_coef$`JAK-STAT`
  progeny_coef$`JAK-STAT` <- NULL
  #lets link the progeny scores to the result object:
  progeny_res <-
    merge(progeny_coef,
          as.data.frame(res),
          by = 0,
          all = TRUE)
  rownames(progeny_res) <- progeny_res$Row.names
  output_filename <-
    paste(comparison_combi_dir, '/progeny_genes.pdf', sep = "")
  
  
  pdf(
    output_filename ,
    width = 8,
    height = 8,
    paper = 'special'
  )
  for (pathway in colnames(progeny_coef)) {
    print(pathway)
    point_plot <- NULL
    subset_Df <- drop_na(progeny_res[c(pathway, 'log2FoldChange')])
    subset_Df <- subset_Df[subset_Df[pathway] != 0,]
    df <-
      gather(subset_Df, category, value, c(-pathway, -'log2FoldChange'))
    df <- mutate_all(df, function(x)
      as.numeric(as.character(x)))
    df$Gene <- rownames(df)
    point_plot <-
      ggplot(df, aes_string(x = colnames(df)[1], y = colnames(df)[2])) +
      geom_point(size = 3)  +
      scale_color_discrete(name = NULL) +
      theme(legend.position = "top")
    point_plot <- point_plot +   geom_label_repel(
      aes(label = Gene),
      box.padding   = 0.35,
      point.padding = 0.5,
      segment.color = 'grey50'
    )
    print(point_plot)
  }
  dev.off()
}

```


Lets load the DEGs of PAX6 knockdown and the DEGS of FOSL2 knockdown
```{r}
FOSL2_DEGS <- read.table('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/FOSL2_knockdown/analysis_output/2023-05-15_costum_backgroun/shCTR_negFC vs shFOSL2_posFC/DEGs.csv', sep = ';')
PAX6_DEGS <- read.table('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/FOSL2_knockdown/analysis_output/2023-05-15_costum_backgroun/shCTR_negFC vs shPAX6_posFC/DEGs.csv', sep = ';')


FOSL2_up <- rownames(FOSL2_DEGS[FOSL2_DEGS$log2FoldChange < 0,])
FOSL2_dwn <- rownames(FOSL2_DEGS[FOSL2_DEGS$log2FoldChange > 0,])
PAX6_up <- rownames(PAX6_DEGS[PAX6_DEGS$log2FoldChange < 0,])
PAX6_dwn <- rownames(PAX6_DEGS[PAX6_DEGS$log2FoldChange > 0,])

listInput <-list(FOSL2up = FOSL2_up, FOSL2dwn = FOSL2_dwn, PAX6up = PAX6_up, PAX6dwn = PAX6_dwn)

library(UpSetR)
upset(fromList(listInput), order.by = "freq")

```
```{r}
FOSL2_up[FOSL2_up %in% PAX6_up]
FOSL2_dwn[FOSL2_dwn %in% PAX6_up]
FOSL2_up[FOSL2_up %in% PAX6_dwn]
FOSL2_dwn[FOSL2_dwn %in% PAX6_dwn]
```



```{r}
getModel(organism = 'Human', top = 200)
```


lets make a heatmap combining the FOSL2 and target genes
```{r}
PAX6_intensities <- read.csv2('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/analysis/PAX6_chipseq/2021-22-03/PAX6_intensity.csv' sep = ';')

```
Load KC/LSC RNAseq matrix
```{r}
mat_KC_LSC <- read.csv2('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/analysis/deseq2/output_figures_deseq2/2022-03-02/LSC_KC/KCvsLSC/all_genes.csv')

# RNA_heatmap <- ComplexHeatmap::Heatmap(mat_KC_LSC,
#   cluster_columns = T,
#   cluster_rows = T,
#   show_row_names = rowname_var,
#   #top_annotation = ha,
#   row_km = 2
#   )
# RNA_heatmap
```


```{r}
mat <- assay(rlog)[row.names(diff_genes), ]
colnames(mat)
mat <- mat[,colnames(mat) %in% c('shCTR_513','shCTR_1094','shCTR_518','shCTR_1099','shFOSL2_513','shFOSL2_1094','shFOSL2_518','shFOSL2_1099')]
mat <- t(scale(t(mat)))

mat_KC_LSC <- mat_KC_LSC[rownames(mat_KC_LSC) %in% rownames(mat),]
mat_KC_LSC$log2FoldChange <- as.numeric(mat_KC_LSC$log2FoldChange)
col_fun = colorRamp2(c(-5,0,5),c('orange','white','purple'))
row_ha <- HeatmapAnnotation(Foldchange = mat_KC_LSC$log2FoldChange, col = c(Foldchange =  col_fun), which = 'row')

mat <- mat[rownames(mat) %in% rownames(mat_KC_LSC),]


RNA_heatmap <- ComplexHeatmap::Heatmap(mat,
  cluster_columns = T,
  cluster_rows = T,
  show_row_names = FALSE,
  right_annotation = row_ha,
  row_km = 2
  )

RNA_heatmap
```

