---
title: "RNAseq_Epithelial_fate_KC_LSC"
author: "Jsmits"
date: "March, 2020"
output:
  html_document: default
  pdf_document: default
---
Conda environment used: 'bulk_rna_seq'
```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=12, fig.height=12) 
knitr::opts_knit$set(root.dir = '/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Publications/gene-regulatory-networks-in-epidermal-and-corneal-epithelia/')
```

```{r, load_libraries, message=FALSE, warning=FALSE}

#run first time (no good anaconda packages, sadly):
#BiocManager::install("DO.db", force = TRUE)#update none
#BiocManager::install("colorRamp2", force = TRUE)#update none

# load libraries
library("biomaRt")
library("plyr")
library('tidyverse')
library("dplyr")
library("stringr")#
library("DESeq2")
library("BiocParallel")
library("vsn")
library("ggplot2")
library("RColorBrewer")
library('cowplot')
library("org.Hs.eg.db")
library('ComplexHeatmap')
library('circlize')
library('ggpubr')
library('grid')
library('gridExtra')
library('gtools')
library('ashr')
library('enrichplot')
library("clusterProfiler")
library('enrichplot')
library('parallel')
library('ViSEAGO')
library("pathview")
library("readxl")
library("ggrepel")
library('msigdbr')
library('EnhancedVolcano')
```

```{r}
work_dir <- '/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Publications/gene-regulatory-networks-in-epidermal-and-corneal-epithelia/'

#import costum written functions:
source(paste0(work_dir, 'Rmarkdowns/R_functions/plot_PCs_RNAseq.R'))
source(paste0(work_dir, 'Rmarkdowns/R_functions/plot_dist_mat_RNAseq.R'))
source(paste0(work_dir, 'Rmarkdowns/R_functions/progeny_pathway_enrichment.R'))
source(paste0(work_dir, 'Rmarkdowns/R_functions/GOenrichment_clustprof_JS.R'))
source(paste0(work_dir, 'Rmarkdowns/R_functions/GOenrichment_plot_counts_JS.R'))
source(paste0(work_dir, 'Rmarkdowns/R_functions/round_df.R'))
```

#set figure directory and create it if it does not exist
```{r set_fig_dir}
deseq2_dir <- paste(paste0(work_dir,'/deseq2/'))
figure_dir <- paste(deseq2_dir, Sys.Date(), sep="")

if (file.exists(deseq2_dir)){print(paste(deseq2_dir, 'exists, overwriting files if present'))} else {dir.create(file.path(deseq2_dir))
  print('generating output directory')
}
if (file.exists(figure_dir)){print(paste(figure_dir, 'exists, overwriting files if present'))} else {dir.create(file.path(figure_dir))
  print('generating output directory')
}
```

#Load all the files
```{r}
count_table_seq2science <- paste0(work_dir,'data/RNAseq_files/GRCh38.p13-counts.tsv')
pseudocount_table <- paste0(work_dir,'data/RNAseq_files/pseudobulk.tsv')
seq2science_sample_file <- paste0(work_dir,'seq2science_files/RNAseq/samples.tsv')
gene_id_file <- paste0(work_dir,'seq2science_files/RNAseq/gene_id2name.tsv')
chromXY_genelist_file <- paste0(work_dir,'data/misc/chrXY.txt')
RNAseq_comp_info_file <- paste0(work_dir,'data/RNAseq_files/RNAseq_comparisons_info.csv')
```

#Lets load the count table data and rename the columns to the seq2science descriptive name instead of the GEO number
```{r read_counts_data}
counts <- read.csv(count_table_seq2science, sep = '\t')
row.names(counts) <- counts$gene
counts$gene <- NULL
counts_pseudo <- read.csv(file=pseudocount_table, sep = '\t')
sample_data<- read.csv2(seq2science_sample_file, header = TRUE, sep = '\t',  comment.char = "#")
colnames(counts) <- sample_data$descriptive_name[match(colnames(counts),sample_data$sample)]
```

#Replace the ENSEMBL gene ID with the HUGO gene name, using the by S2S generated "gene_id2name.tsv" file
```{r replace_gene_ID}
ensg_2_HUGO<- read.csv2(gene_id_file, header = FALSE, sep = '\t',  comment.char = "#")
colnames(ensg_2_HUGO) <- c('ensembl_gene_id','hgnc_symbol')
counts$ensembl_gene_id <- rownames(counts)
rownames(counts) <- NULL
counts_hgnc <- merge(counts, ensg_2_HUGO, by="ensembl_gene_id")
counts_hgnc$ensembl_gene_id <- NULL
counts_hgnc <- distinct(counts_hgnc)
counts_hgnc <- counts_hgnc[,1:ncol(counts_hgnc)] %>% group_by(hgnc_symbol) %>% summarise_all(list(sum))
counts_hgnc <- as.data.frame(counts_hgnc)
rownames(counts_hgnc) <- counts_hgnc$hgnc_symbol
counts_hgnc$hgnc_symbol <- NULL
```

#remove the X&Y chrom, mytochondrial reads and rRNA from the counttables:
```{r filter_count table}
chrXY <- read.csv(chromXY_genelist_file)
counts_hgnc_f <- counts_hgnc[!rownames(counts_hgnc) %in% chrXY$Gene.name,]

counts_pseudo <- counts_pseudo[row.names(counts_pseudo) %in% row.names(counts_hgnc),]
counts_hgnc$gene <- row.names(counts_hgnc)
counts_pseudo$gene <- row.names(counts_pseudo)
all_counts <- merge(counts_hgnc, counts_pseudo, by = 'gene', all.x = TRUE, all.y = TRUE)
row.names(all_counts)<- all_counts$gene
all_counts$gene <- NULL
#remove the X&Y chromosome genes
counts_f <- all_counts[!rownames(all_counts) %in% chrXY$Gene.name,]
counts_f <- counts_f[!grepl('^MT-', rownames(counts_f)),]#remove mitochondria genes
counts_f <- counts_f[!grepl('rRNA', rownames(counts_f)),]#remove ribosomal RNA 
counts_f <- counts_f[!grepl('_', rownames(counts_f)),]#remove ribosomal RNA 
write.table(counts_f, file = paste0(work_dir,'data/RNAseq_files/merged_counttables.tsv'), sep = '\t')
```

Read the info regarding all the comparisons that will be performed 
```{r}
deseq_objects_df <- read.csv2(file = RNAseq_comp_info_file, sep=',',stringsAsFactors=FALSE)
```

```{r}
options(MulticoreParam=MulticoreParam(workers=4))
setwd(work_dir)
for(i in seq_len(nrow(deseq_objects_df))) {
  print(deseq_objects_df[i,]$comparison)
  comparison_figure_dir <- paste(figure_dir ,deseq_objects_df[i,]$comparison, sep = '/')
  dir.create(file.path(comparison_figure_dir), showWarnings = FALSE)
  
  #lets first load the count table, if there are multiple perform an merge
  count_table_list <- str_split(deseq_objects_df[i,]$list_of_count_tables, ';')
  if (length(count_table_list[[1]])==1){
    print('read and use single count file')
    counts <- read.csv(file=count_table_list[[1]], sep = '\t')
  }#if only one counttable file
  if (length(count_table_list[[1]])>1){
    print('read and merge multiple count files')
    count_dfs <- list()
    j = 1
    for (countfile in count_table_list[[1]]){
      print(countfile)
      count_dfs[[j]] <- read.csv(file=countfile, sep = '\t')
      j = j + 1
    }
    print('merging counttable')
    counts <- Reduce(merge, lapply(count_dfs, function(x) data.frame(x, rn = row.names(x))))
    rownames(counts) <- counts$rn
    counts$rn <- NULL
  }#if multiple countfiles, merge them
  counts[is.na(counts)] <- 0

  #lets read the sample file data
  sample_data<- read.csv2(deseq_objects_df[i,]$sample_file, header = TRUE, sep = '\t', comment.char = '#')
  counts <- counts[colnames(counts) %in% sample_data$filename]
  counts <- counts[, sample_data$filename]
  
  #lets export the count table:
  output_filename_merged_counts = paste(comparison_figure_dir, "counts.tsv", sep = '/')
  write.table(as.data.frame(counts), file= output_filename_merged_counts, sep = '\t')

  dds <- DESeqDataSetFromMatrix(countData = counts,
                            #generate deseq object, the samples and designs are specified in deseq_objects_df
                            colData =sample_data,
                            design = as.formula(deseq_objects_df[i,]$design))
  keep <- rowSums(counts(dds)) >= 10
  dds <- dds[keep,]
  dds[[deseq_objects_df$annotation_column1[i]]] <- relevel(dds[[deseq_objects_df$annotation_column1[i]]],deseq_objects_df[i,]$relevel_factor)


  #make a list of all genes measured with more than 10 counts to use as a background.
  gene_background_list <- rownames(counts(dds)[rowSums(counts(dds)) > 10,])
  vst_norm <- rlog(dds) #normalize deseq2 intensities
  
  #lets output a cpm matrix
  fpm_mat <- fpm(dds)
  output_filename_fpkm = paste(comparison_figure_dir, "fpkm.tsv", sep = '/')
  write.table(as.data.frame(fpm_mat), file= output_filename_fpkm, sep = '\t')

  #generate distance matrix
  output_filename = paste(comparison_figure_dir ,paste(paste(deseq_objects_df[i,]$comparison, collapse = ''), "_distance.pdf", sep = ''), sep = '/')
  if(!file.exists(output_filename)){
    plot_dist_matrix(vst_norm,
                     output_filename,
                     deseq_objects_df[i,]$annotation_column1,
                     deseq_objects_df[i,]$annotation_column2)}
  
  #plotPCA graph
  print('plot PCAs (before batch correction)')
  output_filename = paste(comparison_figure_dir ,paste(paste(deseq_objects_df[i,]$comparison, collapse = ''), "_pca.pdf", sep = ''), sep = '/')
  if (!file.exists(paste(comparison_figure_dir,"sig_PCs.RData", sep = '/'))){
    ammount_of_sig_PCs <- JS_plotPCA_graph(vst_norm,
                     intgroup = c(deseq_objects_df[i,]$annotation_column1,deseq_objects_df[i,]$annotation_column2),
                     output_dir =comparison_figure_dir,
                     filename= 'PCA_',
                     run_go_enrichment = TRUE,
                     go_term_n_genes = 250,
                     Genenames_keyType = 'SYMBOL',
                     n_PCA_top_genes = 15,
                     ntop = 1000,
                     PC_cutoff = deseq_objects_df[i,]$PCA_cutoff,
                     gene_background = gene_background_list)
    
    
    save(ammount_of_sig_PCs, file=paste(comparison_figure_dir,"sig_PCs.RData", sep = '/'))
    }
  else{load((paste(comparison_figure_dir,"sig_PCs.RData", sep = '/')))}#load sig PCs

  #perform batch correction
  if (deseq_objects_df[i,]$batch_correction == TRUE){
    print('normalizing batch affect using limma')
    mat <- assay(vst_norm)
    mat <- limma::removeBatchEffect(mat, vst_norm[[(deseq_objects_df[i,]$batch_correction_column)]])
    assay(vst_norm) <- mat

    #generate distance matrix
    output_filename = paste(comparison_figure_dir ,paste(paste(deseq_objects_df[i,]$comparison, collapse = ''), "_batch_corrected_distance.pdf", sep = ''), sep = '/')
    if(!file.exists(output_filename)){
      plot_dist_matrix(vst_norm,
                    output_filename,
                    deseq_objects_df[i,]$annotation_column1,
                    deseq_objects_df[i,]$annotation_column2)}
    #plotPCA graph
    print('plot PCAs after batch correction)')
    output_filename = paste(comparison_figure_dir ,paste(paste(deseq_objects_df[i,]$comparison, collapse = ''), "sig_PCs_batch_corrected.RData", sep = ''), sep = '/')
    if (!file.exists(output_filename)){
      ammount_of_sig_PCs <- JS_plotPCA_graph(vst_norm,
                     intgroup = c(deseq_objects_df[i,]$annotation_column1,deseq_objects_df[i,]$annotation_column2),
                     output_dir =comparison_figure_dir,
                     filename= 'PCA_batch_corrected_',
                     run_go_enrichment = TRUE,
                     go_term_n_genes = 250,
                     Genenames_keyType = 'SYMBOL',
                     n_PCA_top_genes = 15,
                     ntop = 1000,
                     PC_cutoff = deseq_objects_df[i,]$PCA_cutoff,
                     gene_background = gene_background_list)
      save(ammount_of_sig_PCs, file=paste(comparison_figure_dir,"sig_PCs_batch_corrected.RData", sep = '/'))
    }
    else{load((paste(comparison_figure_dir,"sig_PCs.RData", sep = '/')))}#load sig PCs
  }
  #vizualize the counts of the vizualisation genes from the metafile
  if(!is_null(deseq_objects_df$heatmap_genes[i])){
    viz_genes <- deseq_objects_df$heatmap_genes[i]
    print('vizualizing counts of annotation genes')
    myplots <- list()

    for (gene in as.vector(strsplit(viz_genes[1],';')[[1]])){
      if (gene %in% rownames(assay(dds))){
        filename <- paste(comparison_figure_dir, paste0(gene, '.pdf'), sep = '/')
        plot_data <- plotCounts(dds, gene=gene, intgroup=deseq_objects_df$annotation_column1[i], returnData = T)
        colnames(plot_data) <- c('count','annotation1')
        myplots[[gene]] <- local({
            gene <- gene
            p1 <- ggplot(plot_data, aes(x=annotation1, y=count, color=annotation1)) +   scale_y_log10() + geom_point(position=position_jitter(width=.1,height=0), size=3) + ggtitle(paste0("plotCounts of ",gene))
        })
      }}
    pdf(paste(comparison_figure_dir,'/viz_gene_counts.pdf',sep="") ,width=25,height=20,paper='special')
    print(ggarrange(plotlist= myplots, ncol = 4, nrow = 4))
    dev.off()
  }
  
  #perform Progeny pathway enrichment testing on all samplves vs the controll samples:
  output_filename <- paste(comparison_figure_dir,'/PROGENy_barplot_analysis.pdf',sep="")
  output_filename2 <- paste(comparison_figure_dir,'/progeny_data.csv',sep="")
  if(!file.exists(output_filename)){
    print('running progeny pathway enrichiment')
    Progeny_enrichment(vst_norm, 
                       annotation_column_dds = deseq_objects_df$annotation_column1[i], 
                       control_annotation =deseq_objects_df$relevel_factor[i], 
                       output_pdf = output_filename, 
                       output_matrix = output_filename2)
  }
  
  #Perform DEG testing
  print('performing DEG analysis')
  all_DEGs <- list()
  sample_combi <- combinations(n = length(unique(dds[[deseq_objects_df$annotation_column1[i]]])), 2, v= as.vector(dds[[deseq_objects_df$annotation_column1[i]]]))
  dds <- DESeq(dds)

  for (row in 1:nrow(sample_combi)) {
    comp <- (paste(sample_combi[row,1],sample_combi[row,2], sep = 'vs'))
    print(comp)

    #make directory to output graphs and files into
    comparison_combi_dir <- paste(comparison_figure_dir ,comp, sep = '/')
    dir.create(file.path(comparison_combi_dir), showWarnings = FALSE)
    
    #Run Deseq2 for all possible comparisons
    coef_name <- paste0(deseq_objects_df$annotation_column1[i],'_', sample_combi[row,1],"_",'vs',"_",sample_combi[row,2])
    res <- results(dds, contrast = (c(deseq_objects_df$annotation_column1[i], sample_combi[row,1], sample_combi[row,2])), name = comp )
    res <- lfcShrink(dds, contrast = (c(deseq_objects_df$annotation_column1[i], sample_combi[row,1], sample_combi[row,2])),type="ashr")
    #res <- lfcShrink(dds, coef = coef_name,type="apeglm")

    write.table(as.data.frame(res), file= paste(comparison_combi_dir,"/all_genes.csv",sep=""), sep = ';')
    diff_genes <- subset(res, padj < as.numeric(deseq_objects_df[i,]$DEG_pval_cutoff))
    diff_genes <- subset(diff_genes, log2FoldChange > as.numeric(deseq_objects_df[i,]$DEG_FC_cutoff) | log2FoldChange < -1*as.numeric(deseq_objects_df[i,]$DEG_FC_cutoff))
    diff_genes <- as.data.frame(diff_genes)
    
    #filter on FC
    all_DEGs <- append(all_DEGs, row.names(diff_genes))
    write.table(as.data.frame(diff_genes), file= paste(comparison_combi_dir,"/DEGs.csv",sep=""), sep = ';', row.names = T)
    
        
    #make a volcano plot, blegh
    vulcano_plot <- EnhancedVolcano(res,
      lab = rownames(res),
      x = 'log2FoldChange',
      y = 'pvalue',
      title = comp,
      pCutoff = as.numeric(deseq_objects_df[i,]$DEG_pval_cutoff),
      FCcutoff = as.numeric(deseq_objects_df[i,]$DEG_FC_cutoff),
      xlab = bquote(~Log[2]~ 'fold change'),

      pointSize = 4.0,
      labSize = 6.0,
      colAlpha = 1,
      legendPosition = 'right',
      legendLabSize = 12,
      legendIconSize = 4.0,
      drawConnectors = TRUE,
      widthConnectors = 0.75
      )
    
    pdf(paste(comparison_combi_dir,'/Differential_genes_vulcano.pdf',sep="") ,width=12,height=12,paper='special') 
    print(vulcano_plot)
    dev.off()
    
    #run GO term enrichment for all DEGS combined
    output_name <- paste0(comparison_combi_dir,'/all_DEGS')
    if(!file.exists(paste0(output_name,'_GO_enrichment.pdf'))){
      gene_list<- rownames(diff_genes)
      #GO_results <- GOenrichment_clustprof_JS(gene_list, gene_background_list, paste0(output_name,'_GO_enrichment.csv'), simplify_value = 0.6)
      GO_results <- GOenrichment_clustprof_JS(gene_list, gene_background_list, paste0(output_name,'_GO_enrichment.csv'), simplify_value = 0.6)
      if ( GO_results !=  'no sig go terms'){ 
        GOenrichment_plot_counts_JS(GO_results[[1]], dds, res, paste0(output_name,'_GO_enrichment.pdf'), topn_GOterms = 15, topn_genes = 12, 'genotype')} 
      GO_results <- GOenrichment_clustprof_JS(gene_list, paste0(output_name,'_GO_enrichment_human_background.csv'), simplify_value = 0.6)
      if ( GO_results !=  'no sig go terms'){ 
        GOenrichment_plot_counts_JS(GO_results[[1]], dds, res, paste0(output_name,'_GO_enrichment_human_background.pdf'), topn_GOterms = 15, topn_genes = 12, 'genotype')} 
    }

    #run GO term enrichment for all DEGS for high in each side of each comparison
    for (high_in in c(sample_combi[row,1],sample_combi[row,2])){
      if (high_in == sample_combi[row,1]){
        gene_list<- rownames(diff_genes[diff_genes$log2FoldChange > as.numeric(deseq_objects_df[i,]$DEG_FC_cutoff),])
      } else{gene_list <- rownames(diff_genes[diff_genes$log2FoldChange < -as.numeric(deseq_objects_df[i,]$DEG_FC_cutoff),])}
      output_name <- paste(comparison_combi_dir,high_in, sep = '/')
      if(!file.exists(paste0(output_name,'_GO_enrichment.pdf'))){
        print(paste0('running GO enrichment for DEGS of ',high_in))
        GO_results <- GOenrichment_clustprof_JS(gene_list, gene_background_list, paste0(output_name,'_GO_enrichment.csv'), simplify_value = 0.7)
        if ( GO_results !=  'no sig go terms'){ 
          GOenrichment_plot_counts_JS(GO_results[[1]], dds, res, paste0(output_name,'_GO_enrichment.pdf'), topn_GOterms = 15, topn_genes = 12, 'genotype')} 
        GO_results <- GOenrichment_clustprof_JS(gene_list,gene_background_list = 'NA', paste0(output_name,'_GO_enrichment_human_background.csv'), simplify_value = 0.7)
        if ( GO_results !=  'no sig go terms'){ 
          GOenrichment_plot_counts_JS(GO_results[[1]], dds, res, paste0(output_name,'_GO_enrichment_human_background.pdf'), topn_GOterms = 15, topn_genes = 12, 'genotype')} 
      }}
    #run GO term enrichment for all DEGS in general

    #generate a heatmap of all differential genes for each comparison
    output_filename <-  paste0(comparison_combi_dir,'/Differential_genes_heatmap.pdf')
    if(!file.exists(output_filename)){
  
      mat <- assay(vst_norm)[row.names(diff_genes), ]
      mat <- mat - rowMeans(mat)
  
      column_ha <- HeatmapAnnotation(annotation_column1 = vst_norm[[deseq_objects_df[i,]$annotation_column1]], annotation_column2 = vst_norm[[deseq_objects_df[i,]$annotation_column2]])
  
      RNA_heatmap <- Heatmap(mat, 
        cluster_columns = T,
        cluster_rows = T,
        show_row_names = deseq_objects_df[i,]$heatmap_row_names,
        top_annotation = column_ha,
        row_km = 2,
        column_km = 2, 
        column_km_repeats = 100
        )
      
      pdf(paste(comparison_combi_dir,'/Differential_genes_heatmap.pdf',sep="") ,width=8,height=8,paper='special') 
      print(RNA_heatmap)
      dev.off()
    }
    
    #generate a heatmap of a genelist if provided for each comparison
    output_filename <-  paste0(comparison_combi_dir,'/subset_genes_heatmap.pdf')
    if (deseq_objects_df[i,]$subset_heatmap){
      print('generating heatmap of provided gene list')
      gene_subset <- str_split(deseq_objects_df[i,]$subset_heatmap_genes, ';')[[1]]
      gene_subset <- gene_subset[gene_subset %in% rownames(assay(vst_norm))]
      mat <- assay(vst_norm)[gene_subset, ]
      mat <- mat - rowMeans(mat)
      mat <- scale(mat)
      #output the values of the heatmap
      write.table(as.data.frame(mat), file= paste(comparison_combi_dir,"/subset_heatmap_matrix.csv",sep=""), sep = ';', row.names = T)
      pvalue =  res[gene_subset,]$padj
      baseMean =  res[gene_subset,]$baseMean
      write.table(as.data.frame(pvalue), file= paste(comparison_combi_dir,"/subset_heatmap_matrix_pval.csv",sep=""), sep = ';', row.names = T)


      is_sig = pvalue < 0.01
      pch = rep("*", length(mat[,1]))
      pch[!is_sig] = NA
      pvalue_col_fun = colorRamp2(c(0, 2, 3), c("green", "white", "red")) 
      
      ha = rowAnnotation(
          pvalue = anno_simple(-log10(pvalue), col = pvalue_col_fun, pch = pch))
      
      lgd_pvalue = Legend(title = "p-value", col_fun = pvalue_col_fun, at = c(0, 1, 2, 3), 
          labels = c("1", "0.1", "0.01", "0.001"))
      lgd_sig = Legend(pch = "*", type = "points", labels = "< 0.01")
      
      RNA_heatmap <- Heatmap(mat, 
        cluster_columns = T,
        cluster_rows = T,
        show_row_names = T,
        row_km = 7,
        column_km = 2, 
        column_km_repeats = 100,
        right_annotation = ha
        )
      pdf(output_filename ,width=20,height=20,paper='special') 
      draw(RNA_heatmap, annotation_legend_list = list(lgd_pvalue, lgd_sig))
      dev.off()
    }
    
    #runn KEGG enirhcment analysis
    print('running KEGG analysis')
    KEGG_dir <- paste0(comparison_combi_dir, '/KEGG')
    if(!file.exists(KEGG_dir)){
      dir.create(file.path(KEGG_dir), showWarnings = FALSE)
      #convert gene IDs, because of course it doesnt accept HUGO
      entrezIDs <- AnnotationDbi::select(org.Hs.eg.db, 
             keys = rownames(res),
             columns = c("ENTREZID", "SYMBOL"),
             keytype = "SYMBOL")
      
      res_entrez <- res
      res_entrez$SYMBOL <- rownames(res)
      res_entrez <- merge(as.data.frame(res_entrez), entrezIDs, by = 'SYMBOL')
      res_entrez <- drop_na(res_entrez)
      
      FC_list_entrez <- res_entrez$log2FoldChange
      names(FC_list_entrez) <- as.character(res_entrez$ENTREZID)
      geneList_entrez <- sort(FC_list_entrez, decreasing = TRUE)
        
      KEGGenrichment <- gseKEGG(geneList     = geneList_entrez,
                     organism     = 'hsa',
                     minGSSize    = 10,
                     pvalueCutoff = 1,
                     verbose      = FALSE)
      KEGGenrichment <- as.data.frame(KEGGenrichment)
      write.table(  KEGGenrichment,row.names = T, col.names= T, file= paste0(KEGG_dir,"/KEGG_results.csv"), sep = ',')
      
      KEGG_df <- KEGGenrichment[c('Description','NES','qvalues')]
      KEGG_df <- round_df(KEGG_df, digits = 7)
      pdf(file = paste0(KEGG_dir,"/KEGG_results.pdf"), height = 12, width = 26)
      grid.table(KEGG_df[1:10,],rows = NULL)
      dev.off()
      
      #lets vizualize all pathways that are significant
      KEGG_sig <- KEGGenrichment[KEGGenrichment$p.adjust < 0.01,]
      setwd(KEGG_dir)
      for(k in 1:nrow(KEGG_sig)) {
          row_Kegg <- KEGG_sig[k,]
          old_filename = paste0(KEGG_dir,'/',row_Kegg$ID,'.pathview.png')
          new_filename = paste0(KEGG_dir,'/',gsub(" ", "_", row_Kegg$Description),'.pathview.png')
          hsa04668 <- pathview(gene.data  = geneList_entrez,
                           pathway.id = row_Kegg$ID,
                           species    = "hsa",
                           limit      = list(gene=quantile(abs(geneList_entrez), 0.95)),
                           low = list(gene = 'orange'),
                           high = list(gnee = 'purple'))
          file.rename(old_filename, new_filename)
      }
      }
    
    #run GSEA on msigdbr, wikipathways and potentially an costum suplied geneset list
    print('running GSEA')
    GSEA_dir <- paste0(comparison_combi_dir, '/GSEA')
    if(!file.exists(GSEA_dir)){
      dir.create(file.path(GSEA_dir), showWarnings = FALSE)
      m_df <- msigdbr(species = "Homo sapiens")
      res_df <- as.data.frame(res)
      res_df <- drop_na(res_df)
      FC_list <- res_df$log2FoldChange
      names(FC_list) <- as.character(rownames(res_df))
      geneList <- sort(FC_list, decreasing = TRUE)
      for (database in unique(m_df$gs_cat)){
        GSEA_output_dir <- paste(GSEA_dir, database, sep='/' )
        dir.create(file.path(GSEA_output_dir), showWarnings = FALSE)
        print(database)
        msig_df <- msigdbr(species = "Homo sapiens", category = database) %>%   dplyr::select(gs_name, gene_symbol)
        GSEA_results <- clusterProfiler::GSEA(geneList, TERM2GENE = msig_df, pvalueCutoff = 0.1)
        write.table(as.data.frame(GSEA_results),row.names = T, col.names= T, file= paste0(paste(GSEA_dir,database,sep = '/'),'_GSEA_res.csv'), sep = ',')
        GSEA_df <-as.data.frame(GSEA_results)
        GSEA_df <- GSEA_df[c('Description','enrichmentScore','qvalues')]
        GSEA_df <- round_df(GSEA_df, digits = 9)
        pdf(file =paste0(paste(GSEA_dir,database,sep = '/'),'_GSEA_res.pdf'), height = 12, width = 26)
        grid.table(GSEA_df[1:10,], rows = NULL)
        dev.off()
        
        
        gsea_counter = 1
        for (sigGSEA in as.data.frame(GSEA_results@result)[1]$ID){
          if (gsea_counter > 10){break}
          GSEA_plot <- gseaplot2(GSEA_results , geneSetID = gsea_counter, pvalue_table = T)
          plotname <- paste0(sigGSEA, '_GSEA_enrichment.pdf')
          pdf(paste(GSEA_output_dir,plotname, sep="/"),width=8,height=6,paper='special')
          print(GSEA_plot)
          gsea_counter = gsea_counter + 1
          dev.off()}
      }
      
      #If a costum GSEA dir is suplied, also run GSEA on there:
      if(!is.na(deseq_objects_df$GSEA_list[i])){
      print('running GSEA on costum gene list dir supplied')
      #Perform Gene Set Enrichment analysis for a specific Gene Set
      GSEA_dir <- paste0(comparison_combi_dir, '/GSEA')
      dir.create(file.path(GSEA_dir), showWarnings = FALSE)
          
      GSEA_list_dir <- deseq_objects_df$GSEA_list[i]
        #'/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Epi_fate_specifiers/analysis/CREs_model/2021-04-08/model_predictions/' 
      all_gene_sets <- list.files(path = GSEA_list_dir)
      
      GSE_targets <- NULL
      for (geneset in all_gene_sets){
        #check if filesize is not 0
        if (file.info(paste(GSEA_list_dir,geneset,sep='/'))$size > 0){
          print(geneset)
          if (is.null(GSE_targets)){
            
            print(geneset)
            geneset_genes <- read.csv2(paste(GSEA_list_dir,geneset,sep='/'), header = F)
              GSE_targets <- data.frame(
              "term" = geneset,
              "gene" = unique(geneset_genes)
              )}
          else{
              geneset_genes <- read.csv2(paste(GSEA_list_dir,geneset,sep='/'), header = F)
              GSE_targets2 <- data.frame(
              "term" = geneset,
              "gene" = unique(geneset_genes))
            GSE_targets <- rbind(GSE_targets, GSE_targets2)
          }
      
      }}
  
      #vizualize significant GSEA enrichments
        gsea_counter = 1
        for (sigGSEA in as.data.frame(GSEA_results@result)[1]$ID){
          if (gsea_counter > 10){break}
          GSEA_plot <- gseaplot2(GSEA_results , geneSetID = gsea_counter, pvalue_table = T)
          plotname <- paste0(sigGSEA, '_GSEA_enrichment.pdf')
          pdf(paste(GSEA_output_dir,plotname, sep="/"),width=8,height=6,paper='special')
          print(GSEA_plot)
          gsea_counter = gsea_counter + 1
          dev.off()}
      }
  #Run GSEA on the comparison
  #lets load the GSEA objects from the GSEA containing DIR
    if(!is.na(deseq_objects_df$GSEA_list[i])){
      print('running GSEA')
      #Perform Gene Set Enrichment analysis for a specific Gene Set
      GSEA_dir <- paste0(comparison_combi_dir, '/GSEA')
      GSSEA_plots <-  paste0(GSEA_dir, '/costum_db')
      dir.create(file.path(GSEA_dir), showWarnings = FALSE)
      dir.create(file.path(GSSEA_plots), showWarnings = FALSE)
      GSEA_list_dir <- deseq_objects_df$GSEA_list[i]
      all_gene_sets <- list.files(path = GSEA_list_dir)
      GSE_targets <- NULL
      for (geneset in all_gene_sets){
        #check if filesize is not 0
        if (file.info(paste(GSEA_list_dir,geneset,sep='/'))$size > 0){
          if (is.null(GSE_targets)){
              geneset_genes <- read.csv2(paste(GSEA_list_dir,geneset,sep='/'), header = F)
              GSE_targets <- data.frame(
              "term" = geneset,
              "gene" = unique(geneset_genes)
              )}
          else{
              geneset_genes <- read.csv2(paste(GSEA_list_dir,geneset,sep='/'), header = F)
              GSE_targets2 <- data.frame(
              "term" = geneset,
              "gene" = unique(geneset_genes))
            GSE_targets <- rbind(GSE_targets, GSE_targets2)
          }
        }}
      
      #lets make a ranked genelist
      res_GSEA <- as.data.frame(res)
      res_GSEA <- drop_na(res_GSEA)
      FC_list <- res_GSEA$log2FoldChange
      names(FC_list) <- as.character(rownames(res_GSEA))
      geneList <- sort(FC_list, decreasing = TRUE)
      
      GSEA_results <- GSEA(geneList, TERM2GENE = GSE_targets, pvalueCutoff = 1, eps = 0, maxGSSize = 10000)
      #export the GSEA result
      write.table(  as.data.frame(GSEA_results),row.names = T, col.names= T, file= paste0(GSEA_dir,"/GSEA_results.csv"), sep = ',')
      GSEA_df <-as.data.frame(GSEA_results)
      GSEA_df <- GSEA_df[c('Description','enrichmentScore','qvalues')]
      GSEA_df <- round_df(GSEA_df, digits = 7)
      pdf(file = paste0(GSEA_dir,"/GSEA_results.pdf"), height = 12, width = 26)
      grid.table(GSEA_df[1:10,], rows = NULL)
      dev.off()
      
                  
      #if the costum GSEA list is bellow 5 length, lets plot all of them in a single graph:
      if (length(as.data.frame(GSEA_results@result)[1]$ID) < 6) {
        print("costum GSEA db < 5 gene sets, plotting all in one graph")
        GSEA_plot <- gseaplot2(GSEA_results , geneSetID = 1:(length(as.data.frame(GSEA_results@result)[1]$ID)), pvalue_table = T)
        plotname <- paste0('all_costum_genesets_enrichment.pdf')
        pdf(paste(GSSEA_plots,plotname, sep="/"),width=8,height=6,paper='special')
        print(GSEA_plot)
        dev.off()}
      
      #also plot the top 10 GSEA seperate
      
      gseares = 1
      for (sigGSEA in as.data.frame(GSEA_results@result)[1]$ID){
        GSEA_plot <- gseaplot2(GSEA_results , geneSetID = gseares, pvalue_table = T)
        plotname <- paste0(sigGSEA, '_GSEA_enrichment.pdf')
        pdf(paste(GSSEA_plots,plotname, sep="/"),width=8,height=6,paper='special') 
        print(GSEA_plot)
        gseares = gseares + 1
        dev.off()}
    }
    }
    
    #perform Progeny pathway enrichment testing:
    output_filename <- paste(comparison_combi_dir,'/PROGENy_barplot_analysis.pdf',sep="")
    output_filename2 <- paste(comparison_combi_dir,'/progeny_data.csv',sep="")
  
    if(!file.exists(output_filename)){
      print('running progeny pathway enrichiment')
      Progeny_enrichment(vst_norm, 
                         annotation_column_dds = deseq_objects_df$annotation_column1[i], 
                         control_annotation = sample_combi[row,1], 
                         comp_annotation = sample_combi[row,2], 
                         output_pdf = output_filename, 
                         output_matrix = output_filename2)
  
  # vizualize how the gene expression log2FC is behaving compared to the progeny coef numbers:
  url <-   "https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-017-02391-6/MediaObjects/41467_2017_2391_MOESM4_ESM.xlsx"
  destfile <- "~/progeny_coef.xlsx"
  download.file(url, destfile)
  progeny_coef <- read_excel(destfile, col_names = T)
  unlink(destfile)
  progeny_coef <- as.data.frame(progeny_coef)
  colnames(progeny_coef) <- progeny_coef[1,]
  progeny_coef <- progeny_coef[-1,]
  rownames(progeny_coef) <- progeny_coef[,1]
  progeny_coef[,1] <- NULL
  progeny_coef <- progeny_coef[apply(progeny_coef[,-1], 1, function(x) !all(x==0)),]
  #lets link the progeny scores to the result object:
  progeny_res <- merge(progeny_coef, as.data.frame(res), by=0, all = TRUE)
  rownames(progeny_res) <- progeny_res$Row.names
  output_filename <- paste(comparison_combi_dir,'/progeny_genes.pdf',sep="")
  pdf(output_filename ,width=8,height=8,paper='special') 
  for (pathway in colnames(progeny_coef)){
    point_plot <- NULL
    subset_Df <- drop_na(progeny_res[c(pathway,'log2FoldChange')])
    subset_Df <- subset_Df[subset_Df[pathway] != 0,]
    df <- gather(subset_Df, category, value, c(-pathway, -'log2FoldChange'))
    df <- mutate_all(df, function(x) as.numeric(as.character(x)))
    df$Gene <- rownames(df)
    point_plot <- ggplot(df, aes_string(x=colnames(df)[1], y= colnames(df)[2])) +
      geom_point(size=3)  +
      scale_color_discrete(name=NULL) +
      theme(legend.position="top")
    point_plot <- point_plot +   geom_label_repel(aes(label = Gene),
                                    box.padding   = 0.35, 
                                    point.padding = 0.5,
                                    segment.color = 'grey50') 
    print(point_plot)
  }
  dev.off()
  }}
  #run deseq, progeny, KEGG and GSEA for all possible sample combinations
  ###generate a Heatmap to vizualize all the DEGs identified in all comparisons combined
  output_filename <- paste(comparison_figure_dir,'/all_DEGS_heatmap.pdf',sep="")
  if(!file.exists(output_filename)){
    print('generating heatmap of all the DEGS' )
    write.table(t(as.data.frame(all_DEGs)),row.names = FALSE, col.names= F, file= paste(comparison_figure_dir,"/all_DEGs.csv",sep=""))
    all_DEGs <- as.data.frame(all_DEGs)
    all_DEGs <- t(all_DEGs)
    #generate a heatmap of all genes that are differential in a comparison
    mat_all <- assay(vst_norm)
    write.table(mat_all,paste0(comparison_figure_dir,'/all_genes_heatmap_matrix.csv'), sep = ',' )

    mat <- assay(vst_norm)[all_DEGs, ]
    mat <- mat - rowMeans(mat)
    mat <- as.data.frame(mat) %>% distinct()
  
    column_ha <- HeatmapAnnotation(annotation_column1 = vst_norm[[deseq_objects_df[i,]$annotation_column1]], annotation_column2 = vst_norm[[deseq_objects_df[i,]$annotation_column2]])
  
    if(is_null(deseq_objects_df$n_clusters[i])){print('didnt specify a kmeans cluster ammount in the metadata file, using 2')
      kmeans_rows <- 2} else{kmeans_rows <- deseq_objects_df$n_clusters[i]}
    
    #add gene row annotation from the metadata file
    if(!is.na(deseq_objects_df$heatmap_genes[i])){
    highlight_genes <- unlist(strsplit(deseq_objects_df$heatmap_genes[i], split=";"))
    gene_rows <- as.numeric(sapply(highlight_genes, function(x) which(rownames(mat)==x)))
    row_ha = rowAnnotation(selected_genes = anno_mark(at = gene_rows, labels = highlight_genes, labels_gp(fontsize = 40)))
    }else{ row_ha = NA}
    
    write.table(mat,paste0(comparison_figure_dir,'/all_DEGS_heatmap_matrix.csv'), sep = ',' )

    RNA_heatmap <- Heatmap(as.matrix(mat), 
      cluster_columns = T,
      cluster_rows = T,
      show_row_names = deseq_objects_df[i,]$heatmap_row_names,
      #show_column_names = T,
      row_names_gp = gpar(fontsize = 6),
      top_annotation = column_ha,
      right_annotation = row_ha,
      row_km = kmeans_rows ,
      row_km_repeats = 100,
      column_km = length(unique(dds[[deseq_objects_df$annotation_column1[i]]])), 
      column_km_repeats = 100
      )
    
    #Run GO-term enrichment on each cluster of the DEG heatmap
    plot_list <- list()
    clusters <- c('')
    cluster_number = 1
    enriched_GOs = FALSE
    for (cluster in row_order(RNA_heatmap)){
      print(paste('running GO enrichment  DEG cluster ',cluster_number,sep = ''))
      gene_list <- row.names(mat[cluster,])
      PC1_ego <- enrichGO(gene = gene_list,
                          universe = gene_background_list,
                          OrgDb         = org.Hs.eg.db,
                          keyType       = "SYMBOL",
                          ont           = "BP",
                          pAdjustMethod = "BH",
                          #pvalueCutoff  = 0.01,
                          qvalueCutoff  = 0.05)
      if(is.null(PC1_ego@result)){
        go_plot1 <- rectGrob(gp=gpar(col=NA))
        }else{
        PC1_ego <- clusterProfiler::simplify(PC1_ego)
        go_plot1 <- dotplot(PC1_ego, showCategory=20)
        enriched_GOs = TRUE
        }
      plot_list <- c(plot_list, list(NA))
      clusters <- append(clusters, (paste('cluster', cluster_number, sep = '')))
      cluster_number = cluster_number +1
    }
    plot_list <- c(plot_list, list())
    clusters <- append(clusters, '')
    grob1 = grid.grabExpr(draw(RNA_heatmap, annotation_legend_side = "top"), width = 28, height = 20)
    if (enriched_GOs){grob2 = ggarrange(plotlist = plot_list , ncol =1, labels = clusters, align = 'v')}
    
    pdf(paste(comparison_figure_dir,'/all_DEGS_heatmap.pdf',sep="") ,width=20,height=20,paper='special')
    if (enriched_GOs){print(ggarrange(grob1 ,grob2, ncol =2 , nrow = 1, widths= c(5, 4)))}
    else {print(ggarrange(grob1, ncol =1 , nrow = 1, widths= c(5, 4)))}
    dev.off()}
}
  
```


```{r}

```




